Index: crypto/aes/build.info
===================================================================
--- crypto/aes/build.info	(revision 47415)
+++ crypto/aes/build.info	(working copy)
@@ -21,9 +21,10 @@ IF[{- !$disabled{asm} -}]
 
   $AESASM_mips32=aes_cbc.c aes-mips.S
   $AESDEF_mips32=AES_ASM
-  $AESASM_mips64=$AESASM_mips32
-  $AESDEF_mips64=$AESDEF_mips32
 
+#  $AESASM_mips64=$AESASM_mips32
+#  $AESDEF_mips64=$AESDEF_mips32
+
   $AESASM_s390x=aes-s390x.S
   # aes-390x.S implements AES_ctr32_encrypt and AES_xts_[en|de]crypt
   $AESDEF_s390x=AES_ASM AES_CTR_ASM AES_XTS_ASM
