diff -urpN frr-7.5/bgpd/subdir.am frr.ddwrt/bgpd/subdir.am
--- frr-7.5/bgpd/subdir.am	2020-11-10 10:36:52.000000000 +0700
+++ frr.ddwrt/bgpd/subdir.am	2021-01-02 16:32:19.608717803 +0700
@@ -217,7 +217,7 @@ bgpd_bgpd_rpki_la_LDFLAGS = -avoid-versi
 bgpd_bgpd_rpki_la_LIBADD = $(RTRLIB_LIBS)
 
 bgpd_bgpd_bmp_la_SOURCES = bgpd/bgp_bmp.c
-bgpd_bgpd_bmp_la_LIBADD = lib/libfrrcares.la
+#bgpd_bgpd_bmp_la_LIBADD = lib/libfrrcares.la
 bgpd_bgpd_bmp_la_LDFLAGS = -avoid-version -module -shared -export-dynamic
 
 clippy_scan += \
diff -urpN frr-7.5/configure.ac frr.ddwrt/configure.ac
--- frr-7.5/configure.ac	2020-11-10 10:36:52.000000000 +0700
+++ frr.ddwrt/configure.ac	2021-01-02 16:32:19.318722385 +0700
@@ -438,9 +438,9 @@ if test "$enable_static_bin" = "yes"; th
     AC_MSG_ERROR([The --enable-static-bin option must be combined with --enable-static.])
   fi
 fi
-if test "$enable_shared" != "yes"; then
-  AC_MSG_ERROR([FRR cannot be built with --disable-shared.  If you want statically linked daemons, use --enable-shared --enable-static --enable-static-bin])
-fi
+#if test "$enable_shared" != "yes"; then
+#  AC_MSG_ERROR([FRR cannot be built with --disable-shared.  If you want statically linked daemons, use --enable-shared --enable-static --enable-static-bin])
+#fi
 AC_SUBST([AC_LDFLAGS])
 AM_CONDITIONAL([STATIC_BIN], [test "$enable_static_bin" = "yes"])
 
diff -urpN frr-7.5/lib/libfrr.c frr.ddwrt/lib/libfrr.c
--- frr-7.5/lib/libfrr.c	2020-11-10 10:36:52.000000000 +0700
+++ frr.ddwrt/lib/libfrr.c	2021-01-02 16:32:18.968727905 +0700
@@ -1063,8 +1063,8 @@ void frr_run(struct thread_master *maste
 		snprintf(instanceinfo, sizeof(instanceinfo), "instance %u ",
 			 di->instance);
 
-	zlog_notice("%s %s starting: %svty@%d%s", di->name, FRR_VERSION,
-		    instanceinfo, di->vty_port, di->startinfo);
+//	zlog_notice("%s %s starting: %svty@%d%s", di->name, FRR_VERSION,
+//		    instanceinfo, di->vty_port, di->startinfo);
 
 	if (di->terminal) {
 		nodetach_term = true;
diff -urpN frr-7.5/lib/northbound.h frr.ddwrt/lib/northbound.h
--- frr-7.5/lib/northbound.h	2020-11-10 10:36:52.000000000 +0700
+++ frr.ddwrt/lib/northbound.h	2021-01-02 16:32:19.038726804 +0700
@@ -562,11 +562,7 @@ struct frr_yang_module_info {
 
 		/* Priority - lower priorities are processed first. */
 		uint32_t priority;
-#if defined(__GNUC__) && ((__GNUC__ - 0) < 5) && !defined(__clang__)
 	} nodes[YANG_MODULE_MAX_NODES + 1];
-#else
-	} nodes[];
-#endif
 };
 
 /* Northbound error codes. */
diff -urpN frr-7.5/lib/prefix.h frr.ddwrt/lib/prefix.h
--- frr-7.5/lib/prefix.h	2020-11-10 10:36:52.000000000 +0700
+++ frr.ddwrt/lib/prefix.h	2021-01-02 16:32:19.018727118 +0700
@@ -25,11 +25,7 @@
 #ifdef SUNOS_5
 #include <sys/ethernet.h>
 #else
-#ifdef GNU_LINUX
-#include <net/ethernet.h>
-#else
-#include <netinet/if_ether.h>
-#endif
+//#include <linux/if_ether.h>
 #endif
 #include "sockunion.h"
 #include "ipaddr.h"
diff -urpN frr-7.5/lib/zlog.c frr.ddwrt/lib/zlog.c
--- frr-7.5/lib/zlog.c	2020-11-10 10:36:52.000000000 +0700
+++ frr.ddwrt/lib/zlog.c	2021-01-02 16:32:19.028726961 +0700
@@ -448,6 +448,7 @@ static void vzlog_tls(struct zlog_tls *z
 
 void vzlog(int prio, const char *fmt, va_list ap)
 {
+	return;
 	struct zlog_tls *zlog_tls = zlog_tls_get();
 
 	if (zlog_tls)
diff -urpN frr-7.5/nhrpd/subdir.am frr.ddwrt/nhrpd/subdir.am
--- frr-7.5/nhrpd/subdir.am	2020-11-10 10:36:52.000000000 +0700
+++ frr.ddwrt/nhrpd/subdir.am	2021-01-02 16:32:19.128725383 +0700
@@ -8,7 +8,7 @@ vtysh_scan += nhrpd/nhrp_vty.c
 man8 += $(MANBUILD)/frr-nhrpd.8
 endif
 
-nhrpd_nhrpd_LDADD = lib/libfrr.la lib/libfrrcares.la $(LIBCAP)
+nhrpd_nhrpd_LDADD = lib/libfrr.la $(LIBCAP)
 nhrpd_nhrpd_SOURCES = \
 	nhrpd/linux.c \
 	nhrpd/netlink_arp.c \
diff -urpN frr-7.5/tools/frr@.service frr.ddwrt/tools/frr@.service
--- frr-7.5/tools/frr@.service	2020-11-10 10:36:52.000000000 +0700
+++ frr.ddwrt/tools/frr@.service	1970-01-01 07:00:00.000000000 +0700
@@ -1,25 +0,0 @@
-[Unit]
-Description=FRRouting
-Documentation=https://frrouting.readthedocs.io/en/latest/setup.html
-Wants=network.target
-After=network-pre.target systemd-sysctl.service
-Before=network.target
-OnFailure=heartbeat-failed@%n.service
-
-[Service]
-Nice=-5
-Type=forking
-NotifyAccess=all
-StartLimitInterval=3m
-StartLimitBurst=3
-TimeoutSec=2m
-WatchdogSec=60s
-RestartSec=5
-Restart=on-abnormal
-LimitNOFILE=1024
-ExecStart=/usr/lib/frr/frrinit.sh start %I
-ExecStop=/usr/lib/frr/frrinit.sh stop %I
-ExecReload=/usr/lib/frr/frrinit.sh reload %I
-
-[Install]
-WantedBy=multi-user.target
diff -urpN frr-7.5/vtysh/vtysh.c frr.ddwrt/vtysh/vtysh.c
--- frr-7.5/vtysh/vtysh.c	2020-11-10 10:36:52.000000000 +0700
+++ frr.ddwrt/vtysh/vtysh.c	2021-01-02 16:32:19.158724906 +0700
@@ -3096,6 +3096,7 @@ DEFUN (vtysh_write_memory,
 	for (i = 0; i < array_size(vtysh_client); i++)
 		ret = vtysh_client_execute(&vtysh_client[i], line);
 
+	system("startservice quagga_writememory -f");
 	return ret;
 }
 
diff -urpN frr-7.5/zebra/if_netlink.c frr.ddwrt/zebra/if_netlink.c
--- frr-7.5/zebra/if_netlink.c	2020-11-10 10:36:52.000000000 +0700
+++ frr.ddwrt/zebra/if_netlink.c	2021-01-02 16:32:19.318722385 +0700
@@ -30,10 +30,11 @@
  */
 #define _LINUX_IN6_H
 
-#include <netinet/if_ether.h>
+//#include <net/if_ether.h>
 #include <linux/if_bridge.h>
 #include <linux/if_link.h>
 #include <net/if_arp.h>
+//#include <net/ethernet.h>
 #include <linux/sockios.h>
 #include <linux/ethtool.h>
 
