Index: i18n/double-conversion-utils.h
===================================================================
--- i18n/double-conversion-utils.h	(revision 35710)
+++ i18n/double-conversion-utils.h	(revision 35709)
@@ -77,7 +77,7 @@
 // On Linux,x86 89255e-22 != Div_double(89255.0/1e22)
 // ICU PATCH: Enable ARM builds for Windows with 'defined(_M_ARM)'.
 #if defined(_M_X64) || defined(__x86_64__) || \
-    defined(__ARMEL__) || defined(__ARMEB__) || defined(__avr32__) || defined(_M_ARM) || \
+    defined(__ARMEL__) || defined(__avr32__) || defined(_M_ARM) || \
     defined(__hppa__) || defined(__ia64__) || \
     defined(__mips__) || \
     defined(__powerpc__) || defined(__ppc__) || defined(__ppc64__) || \
