Index: inode.c
===================================================================
--- inode.c	(revision 9834)
+++ inode.c	(working copy)
@@ -50,6 +50,7 @@
 
 #ifdef SQUASHFS_LZMA
 #include "LzmaDecode.h"
+#include "LzmaDecode.c"
 
 /* default LZMA settings, should be in sync with mksquashfs */
 #define LZMA_LC 3
Index: LzmaDecode.c
===================================================================
--- LzmaDecode.c	(revision 9834)
+++ LzmaDecode.c	(working copy)
@@ -388,7 +388,7 @@
   return LZMA_RESULT_OK;
 }
 
-int LzmaDecode(unsigned char *buffer,
+static int LzmaDecode(unsigned char *buffer,
     unsigned char *outStream, UInt32 outSize,
     UInt32 *outSizeProcessed)
 {
Index: Makefile
===================================================================
--- Makefile	(revision 9834)
+++ Makefile	(working copy)
@@ -4,5 +4,4 @@
 
 obj-$(CONFIG_SQUASHFS) += squashfs.o
 squashfs-y += inode.o
-squashfs-y += LzmaDecode.o
 CFLAGS_inode.o		:= -O2
