Index: lib/libuutil/Makefile.am
===================================================================
--- lib/libuutil/Makefile.am	(revision 47110)
+++ lib/libuutil/Makefile.am	(working copy)
@@ -21,7 +21,7 @@ libuutil_la_LIBADD = \
 
 libuutil_la_LIBADD += $(LTLIBINTL)
 
-libuutil_la_LDFLAGS = -pthread
+libuutil_la_LDFLAGS = -pthread -latomic
 
 if !ASAN_ENABLED
 libuutil_la_LDFLAGS += -Wl,-z,defs
Index: lib/libzfs_core/Makefile.am
===================================================================
--- lib/libzfs_core/Makefile.am	(revision 47110)
+++ lib/libzfs_core/Makefile.am	(working copy)
@@ -19,7 +19,7 @@ libzfs_core_la_LIBADD = \
 
 libzfs_core_la_LIBADD += $(LTLIBINTL)
 
-libzfs_core_la_LDFLAGS = -pthread
+libzfs_core_la_LDFLAGS = -pthread -latomic
 
 if !ASAN_ENABLED
 libzfs_core_la_LDFLAGS += -Wl,-z,defs
