Index: tests/zfs-tests/cmd/mmap_seek/mmap_seek.c
===================================================================
--- tests/zfs-tests/cmd/mmap_seek/mmap_seek.c	(revision 47709)
+++ tests/zfs-tests/cmd/mmap_seek/mmap_seek.c	(working copy)
@@ -31,6 +31,18 @@
 #include <sys/mman.h>
 #include <errno.h>
 
+#ifndef SEEK_DATA
+#define SEEK_DATA 3
+#endif
+
+#ifndef SEEK_HOLE
+#define SEEK_HOLE 4
+#endif
+
+#ifndef P2ROUNDUP
+#define	P2ROUNDUP(x, align)	((((x) - 1) | ((align) - 1)) + 1)
+#endif
+
 static void
 seek_data(int fd, off_t offset, off_t expected)
 {
