Index: module/zcommon/zfs_fletcher.c
===================================================================
--- module/zcommon/zfs_fletcher.c	(revision 52639)
+++ module/zcommon/zfs_fletcher.c	(working copy)
@@ -188,7 +188,7 @@ static const fletcher_4_ops_t *fletcher_4_impls[]
 #if defined(__x86_64) && defined(HAVE_AVX512BW)
 	&fletcher_4_avx512bw_ops,
 #endif
-#if defined(__aarch64__) && !defined(__FreeBSD__)
+#if 0 //defined(__aarch64__) && !defined(__FreeBSD__)
 	&fletcher_4_aarch64_neon_ops,
 #endif
 };
Index: module/zcommon/zfs_fletcher_aarch64_neon.c
===================================================================
--- module/zcommon/zfs_fletcher_aarch64_neon.c	(revision 52639)
+++ module/zcommon/zfs_fletcher_aarch64_neon.c	(working copy)
@@ -41,7 +41,7 @@
  * SOFTWARE.
  */
 
-#if defined(__aarch64__)
+#if 0 //defined(__aarch64__)
 
 #include <sys/simd.h>
 #include <sys/spa_checksum.h>
