Index: include/os/linux/kernel/linux/vfs_compat.h
===================================================================
--- include/os/linux/kernel/linux/vfs_compat.h	(revision 48909)
+++ include/os/linux/kernel/linux/vfs_compat.h	(working copy)
@@ -146,6 +146,11 @@ zpl_bdi_destroy(struct super_block *sb)
 #define	clear_inode(ip)		end_writeback(ip)
 #endif /* HAVE_EVICT_INODE && !HAVE_CLEAR_INODE */
 
+#ifndef SEEK_DATA
+#define SEEK_DATA 3
+#define SEEK_HOLE 4
+#endif
+
 #if defined(SEEK_HOLE) && defined(SEEK_DATA) && !defined(HAVE_LSEEK_EXECUTE)
 static inline loff_t
 lseek_execute(
Index: include/os/linux/spl/sys/vnode.h
===================================================================
--- include/os/linux/spl/sys/vnode.h	(revision 48909)
+++ include/os/linux/spl/sys/vnode.h	(working copy)
@@ -52,6 +52,10 @@
 
 #define	F_FREESP	11 	/* Free file space */
 
+#ifndef SEEK_DATA
+#define SEEK_DATA 3
+#define SEEK_HOLE 4
+#endif
 
 #if defined(SEEK_HOLE) && defined(SEEK_DATA)
 #define	F_SEEK_DATA	SEEK_DATA
Index: module/os/linux/zfs/zpl_file.c
===================================================================
--- module/os/linux/zfs/zpl_file.c	(revision 48909)
+++ module/os/linux/zfs/zpl_file.c	(working copy)
@@ -545,6 +545,11 @@ zpl_direct_IO(int rw, struct kiocb *kiocb, struct
 
 #endif /* HAVE_VFS_RW_ITERATE */
 
+#ifndef SEEK_DATA
+#define SEEK_DATA 3
+#define SEEK_HOLE 4
+#endif
+
 static loff_t
 zpl_llseek(struct file *filp, loff_t offset, int whence)
 {
Index: module/zfs/zfs_vnops.c
===================================================================
--- module/zfs/zfs_vnops.c	(revision 48909)
+++ module/zfs/zfs_vnops.c	(working copy)
@@ -78,6 +78,10 @@ zfs_fsync(znode_t *zp, int syncflag, cred_t *cr)
 	return (0);
 }
 
+#ifndef SEEK_DATA
+#define SEEK_DATA 3
+#define SEEK_HOLE 4
+#endif
 
 #if defined(SEEK_HOLE) && defined(SEEK_DATA)
 /*
Index: tests/zfs-tests/cmd/mmap_seek.c
===================================================================
--- tests/zfs-tests/cmd/mmap_seek.c	(revision 48909)
+++ tests/zfs-tests/cmd/mmap_seek.c	(working copy)
@@ -35,6 +35,11 @@
 #include <linux/fs.h>
 #endif
 
+#ifndef SEEK_DATA
+#define SEEK_DATA 3
+#define SEEK_HOLE 4
+#endif
+
 static void
 seek_data(int fd, off_t offset, off_t expected)
 {
