Index: etc/systemd/system-generators/zfs-mount-generator.c
===================================================================
--- etc/systemd/system-generators/zfs-mount-generator.c	(revision 50434)
+++ etc/systemd/system-generators/zfs-mount-generator.c	(working copy)
@@ -193,7 +193,7 @@ fopenat(int dirfd, const char *pathname, int flags
 	return (fdopen(fd, stream_mode));
 }
 
-#define	my_strndupa(x,y)	strncpy(alloca(y),x,y)
+#define	my_strndupa(x,y)	strlcpy(alloca(y),x,y)
 
 static int
 line_worker(char *line, const char *cachefile)
Index: tests/zfs-tests/cmd/readmmap.c
===================================================================
--- tests/zfs-tests/cmd/readmmap.c	(revision 50434)
+++ tests/zfs-tests/cmd/readmmap.c	(working copy)
@@ -45,6 +45,7 @@
 #include <errno.h>
 #include <sys/mman.h>
 #include <time.h>
+#include <sys/types.h>
 
 int
 main(int argc, char **argv)
