Index: lib/libzfs/libzfs_crypto.c
===================================================================
--- lib/libzfs/libzfs_crypto.c	(revision 46727)
+++ lib/libzfs/libzfs_crypto.c	(working copy)
@@ -28,7 +28,6 @@
 #include <openssl/evp.h>
 #if LIBFETCH_DYNAMIC
 #include <dlfcn.h>
-#endif
 #if LIBFETCH_IS_FETCH
 #include <sys/param.h>
 #include <stdio.h>
@@ -36,6 +35,7 @@
 #elif LIBFETCH_IS_LIBCURL
 #include <curl/curl.h>
 #endif
+#endif
 #include <libzfs.h>
 #include "libzfs_impl.h"
 #include "zfeature_common.h"
