Index: cmd/zed/agents/fmd_api.c
===================================================================
--- cmd/zed/agents/fmd_api.c	(revision 59079)
+++ cmd/zed/agents/fmd_api.c	(working copy)
@@ -37,7 +37,7 @@
 
 #include <sys/types.h>
 #include <sys/fm/protocol.h>
-#include <uuid/uuid.h>
+#include <uuid.h>
 #include <signal.h>
 #include <string.h>
 #include <time.h>
Index: cmd/zpool/os/linux/zpool_vdev_os.c
===================================================================
--- cmd/zpool/os/linux/zpool_vdev_os.c	(revision 59079)
+++ cmd/zpool/os/linux/zpool_vdev_os.c	(working copy)
@@ -82,8 +82,8 @@
 #include <sys/efi_partition.h>
 #include <sys/stat.h>
 #include <sys/mntent.h>
-#include <uuid/uuid.h>
-#include <blkid/blkid.h>
+#include <uuid.h>
+#include <blkid.h>
 
 typedef struct vdev_disk_db_entry
 {
Index: config/kernel-kuidgid.m4
===================================================================
--- config/kernel-kuidgid.m4	(revision 59079)
+++ config/kernel-kuidgid.m4	(working copy)
@@ -1,21 +1,28 @@
 dnl #
-dnl # 3.8 API change,
-dnl # User namespaces, use kuid_t in place of uid_t where available.
+dnl # User namespaces, use kuid_t in place of uid_t
+dnl # where available. Not strictly a user namespaces thing
+dnl # but it should prevent surprises
 dnl #
-AC_DEFUN([ZFS_AC_KERNEL_SRC_KUIDGID_T], [
-	ZFS_LINUX_TEST_SRC([kuidgid_t], [
+AC_DEFUN([ZFS_AC_KERNEL_KUIDGID_T], [
+	AC_MSG_CHECKING([whether kuid_t/kgid_t is available])
+	ZFS_LINUX_TRY_COMPILE([
 		#include <linux/uidgid.h>
 	], [
 		kuid_t userid __attribute__ ((unused)) = KUIDT_INIT(0);
 		kgid_t groupid __attribute__ ((unused)) = KGIDT_INIT(0);
-	])
-])
-
-AC_DEFUN([ZFS_AC_KERNEL_KUIDGID_T], [
-	AC_MSG_CHECKING([whether kuid_t/kgid_t is available])
-	ZFS_LINUX_TEST_RESULT([kuidgid_t], [
-		AC_MSG_RESULT(yes)
 	],[
-		ZFS_LINUX_TEST_ERROR([kuid_t/kgid_t])
+		ZFS_LINUX_TRY_COMPILE([
+			#include <linux/uidgid.h>
+		], [
+			kuid_t userid __attribute__ ((unused)) = 0;
+			kgid_t groupid __attribute__ ((unused)) = 0;
+		],[
+			AC_MSG_RESULT(yes; optional)
+		],[
+			AC_MSG_RESULT(yes; mandatory)
+			AC_DEFINE(HAVE_KUIDGID_T, 1, [kuid_t/kgid_t in use])
+		])
+	],[
+		AC_MSG_RESULT(no)
 	])
 ])
Index: config/kernel.m4
===================================================================
--- config/kernel.m4	(revision 59079)
+++ config/kernel.m4	(working copy)
@@ -94,7 +94,6 @@ AC_DEFUN([ZFS_AC_KERNEL_TEST_SRC], [
 	ZFS_AC_KERNEL_SRC_GENERIC_IO_ACCT
 	ZFS_AC_KERNEL_SRC_FPU
 	ZFS_AC_KERNEL_SRC_FMODE_T
-	ZFS_AC_KERNEL_SRC_KUIDGID_T
 	ZFS_AC_KERNEL_SRC_KUID_HELPERS
 	ZFS_AC_KERNEL_SRC_RENAME
 	ZFS_AC_KERNEL_SRC_TOTALRAM_PAGES_FUNC
@@ -688,7 +687,7 @@ AC_DEFUN([ZFS_LINUX_COMPILE], [
 	    KBUILD_MODPOST_NOFINAL="$5" KBUILD_MODPOST_WARN="$6"
 	    make modules -k -j$TEST_JOBS ${KERNEL_CC:+CC=$KERNEL_CC}
 	    ${KERNEL_LD:+LD=$KERNEL_LD} ${KERNEL_LLVM:+LLVM=$KERNEL_LLVM}
-	    CONFIG_MODULES=y CFLAGS_MODULE=-DCONFIG_MODULES
+	    CONFIG_MODULES=y CFLAGS_MODULE="-DCONFIG_MODULES -fno-lto"
 	    ${KERNEL_CROSS_COMPILE:+CROSS_COMPILE=$KERNEL_CROSS_COMPILE}
 	    ${KERNEL_ARCH:+ARCH=$KERNEL_ARCH}
 	    -C $LINUX_OBJ $ARCH_UM M=$PWD/$1 >$1/build.log 2>&1])
Index: config/user-libblkid.m4
===================================================================
--- config/user-libblkid.m4	(revision 59079)
+++ config/user-libblkid.m4	(working copy)
@@ -3,7 +3,11 @@ dnl # Check for libblkid.  Basic support for detec
 dnl # has existing in blkid since 2008.
 dnl #
 AC_DEFUN([ZFS_AC_CONFIG_USER_LIBBLKID], [
-	ZFS_AC_FIND_SYSTEM_LIBRARY(LIBBLKID, [blkid], [blkid/blkid.h], [], [blkid], [], [], [
-		AC_MSG_FAILURE([
-		*** blkid.h missing, libblkid-devel package required])])
+	LIBBLKID=
+
+	AC_CHECK_HEADER([blkid.h], [], [AC_MSG_FAILURE([
+	*** blkid.h missing, libblkid-devel package required])])
+
+	AC_SUBST([LIBBLKID_LIBS], ["-lblkid"])
+	AC_DEFINE([HAVE_LIBBLKID], 1, [Define if you have libblkid])
 ])
Index: config/user-libuuid.m4
===================================================================
--- config/user-libuuid.m4	(revision 59079)
+++ config/user-libuuid.m4	(working copy)
@@ -2,7 +2,17 @@ dnl #
 dnl # Check for libuuid
 dnl #
 AC_DEFUN([ZFS_AC_CONFIG_USER_LIBUUID], [
-	ZFS_AC_FIND_SYSTEM_LIBRARY(LIBUUID, [uuid], [uuid/uuid.h], [], [uuid], [uuid_generate uuid_is_null], [], [
-	    AC_MSG_FAILURE([*** libuuid-devel package required])
-	])
+	LIBUUID=
+
+	AC_CHECK_HEADER([uuid.h], [], [AC_MSG_FAILURE([
+	*** uuid/uuid.h missing, libuuid-devel package required])])
+
+	AC_SEARCH_LIBS([uuid_generate], [uuid], [], [AC_MSG_FAILURE([
+	*** uuid_generate() missing, libuuid-devel package required])])
+
+	AC_SEARCH_LIBS([uuid_is_null], [uuid], [], [AC_MSG_FAILURE([
+	*** uuid_is_null() missing, libuuid-devel package required])])
+
+	AC_SUBST([LIBUUID], ["-luuid"])
+	AC_DEFINE([HAVE_LIBUUID], 1, [Define if you have libuuid])
 ])
Index: include/os/linux/kernel/linux/vfs_compat.h
===================================================================
--- include/os/linux/kernel/linux/vfs_compat.h	(revision 59079)
+++ include/os/linux/kernel/linux/vfs_compat.h	(working copy)
@@ -60,6 +60,11 @@
 #define	SB_NOATIME	MS_NOATIME
 #endif
 
+#ifndef SEEK_DATA
+#define SEEK_DATA 3
+#define SEEK_HOLE 4
+#endif
+
 #if defined(SEEK_HOLE) && defined(SEEK_DATA)
 static inline loff_t
 lseek_execute(
Index: include/os/linux/spl/sys/cred.h
===================================================================
--- include/os/linux/spl/sys/cred.h	(revision 59079)
+++ include/os/linux/spl/sys/cred.h	(working copy)
@@ -42,6 +42,10 @@ extern struct task_struct init_task;
 #define	GROUP_AT(gi, i)	((gi)->gid[i])
 #endif
 
+extern zidmap_t *zfs_get_init_idmap(void);
+
+#ifdef HAVE_KUIDGID_T
+
 #define	KUID_TO_SUID(x)		(__kuid_val(x))
 #define	KGID_TO_SGID(x)		(__kgid_val(x))
 #define	SUID_TO_KUID(x)		(KUIDT_INIT(x))
@@ -48,8 +52,17 @@ extern struct task_struct init_task;
 #define	SGID_TO_KGID(x)		(KGIDT_INIT(x))
 #define	KGIDP_TO_SGIDP(x)	(&(x)->val)
 
-extern zidmap_t *zfs_get_init_idmap(void);
 
+#else /* HAVE_KUIDGID_T */
+
+#define	KUID_TO_SUID(x)		(x)
+#define	KGID_TO_SGID(x)		(x)
+#define	SUID_TO_KUID(x)		(x)
+#define	SGID_TO_KGID(x)		(x)
+#define	KGIDP_TO_SGIDP(x)	(x)
+
+#endif /* HAVE_KUIDGID_T */
+
 /* Check if the user ns is the initial one */
 static inline boolean_t
 zfs_is_init_userns(struct user_namespace *user_ns)
@@ -171,6 +184,7 @@ static inline gid_t zfs_vfsgid_to_gid(zidmap_t *mn
 	return (__kgid_val(make_kgid(fs_userns, gid)));
 }
 
+
 extern void crhold(cred_t *cr);
 extern void crfree(cred_t *cr);
 extern uid_t crgetuid(const cred_t *cr);
Index: include/os/linux/spl/sys/mutex.h
===================================================================
--- include/os/linux/spl/sys/mutex.h	(revision 59079)
+++ include/os/linux/spl/sys/mutex.h	(working copy)
@@ -27,6 +27,7 @@
 #include <sys/types.h>
 #include <linux/sched.h>
 #include <linux/mutex.h>
+#include <linux/sched.h>
 #include <linux/lockdep.h>
 #include <linux/compiler_compat.h>
 
Index: include/os/linux/spl/sys/vnode.h
===================================================================
--- include/os/linux/spl/sys/vnode.h	(revision 59079)
+++ include/os/linux/spl/sys/vnode.h	(working copy)
@@ -52,6 +52,10 @@
 
 #define	F_FREESP	11 	/* Free file space */
 
+#ifndef SEEK_DATA
+#define SEEK_DATA 3
+#define SEEK_HOLE 4
+#endif
 
 #if defined(SEEK_HOLE) && defined(SEEK_DATA)
 #define	F_SEEK_DATA	SEEK_DATA
Index: lib/libefi/rdwr_efi.c
===================================================================
--- lib/libefi/rdwr_efi.c	(revision 59079)
+++ lib/libefi/rdwr_efi.c	(working copy)
@@ -30,7 +30,7 @@
 #include <errno.h>
 #include <string.h>
 #include <unistd.h>
-#include <uuid/uuid.h>
+#include <uuid.h>
 #include <zlib.h>
 #include <libintl.h>
 #include <sys/types.h>
Index: lib/libshare/os/linux/nfs.c
===================================================================
--- lib/libshare/os/linux/nfs.c	(revision 59079)
+++ lib/libshare/os/linux/nfs.c	(working copy)
@@ -29,6 +29,7 @@
 #include <dirent.h>
 #include <stdio.h>
 #include <string.h>
+#include <fcntl.h>
 #include <errno.h>
 #include <fcntl.h>
 #include <sys/file.h>
Index: lib/libspl/include/sys/backtrace.h
===================================================================
--- lib/libspl/include/sys/backtrace.h	(revision 59079)
+++ lib/libspl/include/sys/backtrace.h	(working copy)
@@ -27,6 +27,10 @@
 #ifndef _LIBSPL_SYS_BACKTRACE_H
 #define	_LIBSPL_SYS_BACKTRACE_H
 
+#ifndef __maybe_unused
+#define	__maybe_unused __attribute__((unused))
+#endif
+
 void libspl_backtrace(int fd);
 
 #endif
Index: lib/libuutil/Makefile.am
===================================================================
--- lib/libuutil/Makefile.am	(revision 59079)
+++ lib/libuutil/Makefile.am	(working copy)
@@ -17,7 +17,7 @@ libuutil_la_LIBADD = \
 
 libuutil_la_LIBADD += $(LTLIBINTL)
 
-libuutil_la_LDFLAGS = -pthread
+libuutil_la_LDFLAGS = -pthread -latomic
 
 if !ASAN_ENABLED
 libuutil_la_LDFLAGS += -Wl,-z,defs
Index: lib/libzfs_core/Makefile.am
===================================================================
--- lib/libzfs_core/Makefile.am	(revision 59079)
+++ lib/libzfs_core/Makefile.am	(working copy)
@@ -29,7 +29,7 @@ libzfs_core_la_LIBADD = \
 
 libzfs_core_la_LIBADD += $(LTLIBINTL)
 
-libzfs_core_la_LDFLAGS = -pthread
+libzfs_core_la_LDFLAGS = -pthread -latomic
 
 if !ASAN_ENABLED
 libzfs_core_la_LDFLAGS += -Wl,-z,defs
Index: lib/libzpool/Makefile.am
===================================================================
--- lib/libzpool/Makefile.am	(revision 59079)
+++ lib/libzpool/Makefile.am	(working copy)
@@ -202,7 +202,7 @@ libzpool_la_LIBADD = \
 
 libzpool_la_LIBADD += $(LIBCLOCK_GETTIME) $(ZLIB_LIBS) -ldl -lm
 
-libzpool_la_LDFLAGS = -pthread
+libzpool_la_LDFLAGS = -pthread -latomic
 
 if !ASAN_ENABLED
 libzpool_la_LDFLAGS += -Wl,-z,defs
Index: lib/libzutil/os/linux/zutil_import_os.c
===================================================================
--- lib/libzutil/os/linux/zutil_import_os.c	(revision 59079)
+++ lib/libzutil/os/linux/zutil_import_os.c	(working copy)
@@ -73,7 +73,7 @@
 #include <libudev.h>
 #include <sched.h>
 #endif
-#include <blkid/blkid.h>
+#include <blkid.h>
 
 #define	DEV_BYID_PATH	"/dev/disk/by-id/"
 
@@ -580,7 +580,7 @@ zfs_device_get_physical(struct udev_device *dev, c
 int
 zpool_label_disk_wait(const char *path, int timeout_ms)
 {
-#ifdef HAVE_LIBUDEV
+#if 0 //def HAVE_LIBUDEV
 	struct udev *udev;
 	struct udev_device *dev = NULL;
 	char nodepath[MAXPATHLEN];
Index: module/Kbuild.in
===================================================================
--- module/Kbuild.in	(revision 59079)
+++ module/Kbuild.in	(working copy)
@@ -2,7 +2,7 @@
 # first.  This ensures its module initialization function is run before
 # any of the other module initialization functions which depend on it.
 
-ZFS_MODULE_CFLAGS += -std=gnu99 -Wno-declaration-after-statement
+ZFS_MODULE_CFLAGS += -std=gnu99 -Wno-declaration-after-statement -fno-lto
 ZFS_MODULE_CFLAGS += -Wmissing-prototypes
 ZFS_MODULE_CFLAGS += @KERNEL_DEBUG_CFLAGS@  @NO_FORMAT_ZERO_LENGTH@
 
Index: module/os/linux/zfs/zfs_vfsops.c
===================================================================
--- module/os/linux/zfs/zfs_vfsops.c	(revision 59079)
+++ module/os/linux/zfs/zfs_vfsops.c	(working copy)
@@ -1215,7 +1215,7 @@ zfs_prune(struct super_block *sb, unsigned long nr
 	if ((error = zfs_enter(zfsvfs, FTAG)) != 0)
 		return (error);
 
-#ifdef SHRINKER_NUMA_AWARE
+#if defined(SHRINKER_NUMA_AWARE) && defined(SHRINK_EMPTY)
 	if (shrinker->flags & SHRINKER_NUMA_AWARE) {
 		long tc = 1;
 		for_each_online_node(sc.nid) {
Index: module/os/linux/zfs/zpl_file.c
===================================================================
--- module/os/linux/zfs/zpl_file.c	(revision 59079)
+++ module/os/linux/zfs/zpl_file.c	(working copy)
@@ -309,6 +309,11 @@ zpl_direct_IO(struct kiocb *kiocb, struct iov_iter
 	return (0);
 }
 
+#ifndef SEEK_DATA
+#define SEEK_DATA 3
+#define SEEK_HOLE 4
+#endif
+
 static loff_t
 zpl_llseek(struct file *filp, loff_t offset, int whence)
 {
Index: module/zfs/zfs_vnops.c
===================================================================
--- module/zfs/zfs_vnops.c	(revision 59079)
+++ module/zfs/zfs_vnops.c	(working copy)
@@ -110,6 +110,10 @@ zfs_fsync(znode_t *zp, int syncflag, cred_t *cr)
 	return (error);
 }
 
+#ifndef SEEK_DATA
+#define SEEK_DATA 3
+#define SEEK_HOLE 4
+#endif
 
 #if defined(SEEK_HOLE) && defined(SEEK_DATA)
 /*
Index: tests/zfs-tests/cmd/mmap_seek.c
===================================================================
--- tests/zfs-tests/cmd/mmap_seek.c	(revision 59079)
+++ tests/zfs-tests/cmd/mmap_seek.c	(working copy)
@@ -45,6 +45,11 @@
 #endif
 #endif
 
+#ifndef SEEK_DATA
+#define SEEK_DATA 3
+#define SEEK_HOLE 4
+#endif
+
 static void
 seek_data(int fd, off_t offset, off_t expected)
 {
Index: tests/zfs-tests/cmd/readmmap.c
===================================================================
--- tests/zfs-tests/cmd/readmmap.c	(revision 59079)
+++ tests/zfs-tests/cmd/readmmap.c	(working copy)
@@ -46,6 +46,7 @@
 #include <sys/mman.h>
 #include <sys/types.h>
 #include <time.h>
+#include <sys/types.h>
 
 int
 main(int argc, char **argv)
Index: config/kernel-config-defined.m4
===================================================================
--- config/kernel-config-defined.m4	(revision 62830)
+++ config/kernel-config-defined.m4	(working copy)
@@ -7,7 +7,6 @@ AC_DEFUN([ZFS_AC_KERNEL_CONFIG_DEFINED], [
 	ZFS_AC_KERNEL_SRC_CONFIG_MODULES
 	ZFS_AC_KERNEL_SRC_CONFIG_BLOCK
 	ZFS_AC_KERNEL_SRC_CONFIG_DEBUG_LOCK_ALLOC
-	ZFS_AC_KERNEL_SRC_CONFIG_TRIM_UNUSED_KSYMS
 	ZFS_AC_KERNEL_SRC_CONFIG_ZLIB_DEFLATE
 	ZFS_AC_KERNEL_SRC_CONFIG_ZLIB_INFLATE
 
@@ -18,7 +17,6 @@ AC_DEFUN([ZFS_AC_KERNEL_CONFIG_DEFINED], [
 	ZFS_AC_KERNEL_CONFIG_MODULES
 	ZFS_AC_KERNEL_CONFIG_BLOCK
 	ZFS_AC_KERNEL_CONFIG_DEBUG_LOCK_ALLOC
-	ZFS_AC_KERNEL_CONFIG_TRIM_UNUSED_KSYMS
 	ZFS_AC_KERNEL_CONFIG_ZLIB_DEFLATE
 	ZFS_AC_KERNEL_CONFIG_ZLIB_INFLATE
 ])
@@ -142,33 +140,6 @@ AC_DEFUN([ZFS_AC_KERNEL_CONFIG_MODULES], [
 ])
 
 dnl #
-dnl # Check CONFIG_TRIM_UNUSED_KSYMS
-dnl #
-dnl # Verify the kernel has CONFIG_TRIM_UNUSED_KSYMS disabled.
-dnl #
-AC_DEFUN([ZFS_AC_KERNEL_SRC_CONFIG_TRIM_UNUSED_KSYMS], [
-	ZFS_LINUX_TEST_SRC([config_trim_unusued_ksyms], [
-		#if defined(CONFIG_TRIM_UNUSED_KSYMS)
-		#error CONFIG_TRIM_UNUSED_KSYMS not defined
-		#endif
-	],[])
-])
-
-AC_DEFUN([ZFS_AC_KERNEL_CONFIG_TRIM_UNUSED_KSYMS], [
-	AC_MSG_CHECKING([whether CONFIG_TRIM_UNUSED_KSYM is disabled])
-	ZFS_LINUX_TEST_RESULT([config_trim_unusued_ksyms], [
-		AC_MSG_RESULT([yes])
-	],[
-		AC_MSG_RESULT([no])
-		AS_IF([test "x$enable_linux_builtin" != xyes], [
-			AC_MSG_ERROR([
-	*** This kernel has unused symbols trimming enabled, please disable.
-	*** Rebuild the kernel with CONFIG_TRIM_UNUSED_KSYMS=n set.])
-		])
-	])
-])
-
-dnl #
 dnl # Check CONFIG_ZLIB_INFLATE
 dnl #
 dnl # Verify the kernel has CONFIG_ZLIB_INFLATE support enabled.
Index: include/os/linux/zfs/sys/zfs_debug_os.h
===================================================================
--- include/os/linux/zfs/sys/zfs_debug_os.h	(revision 62830)
+++ include/os/linux/zfs/sys/zfs_debug_os.h	(working copy)
@@ -23,7 +23,8 @@
 #ifndef _SYS_ZFS_DEBUG_OS_H
 #define	_SYS_ZFS_DEBUG_OS_H
 
-#define	 SET_ERROR(err) \
-	(__set_error(__FILE__, __func__, __LINE__, err), err)
+#define	SET_ERROR(err)	({ 					\
+	err;							\
+})
 
 #endif	/* _SYS_ZFS_DEBUG_OS_H */
Index: lib/libspl/include/sys/kstat.h
===================================================================
--- lib/libspl/include/sys/kstat.h	(revision 62830)
+++ lib/libspl/include/sys/kstat.h	(working copy)
@@ -86,6 +86,6 @@ extern void kstat_delete(kstat_t *);
 extern void kstat_set_raw_ops(kstat_t *ksp,
     int (*headers)(char *buf, size_t size),
     int (*data)(char *buf, size_t size, void *data),
-    void *(*addr)(kstat_t *ksp, loff_t index));
+    void *(*addr)(kstat_t *ksp, off_t index));
 
 #endif	/* _SYS_KSTAT_H */
Index: lib/libspl/kstat.c
===================================================================
--- lib/libspl/kstat.c	(revision 62830)
+++ lib/libspl/kstat.c	(working copy)
@@ -58,7 +58,7 @@ void
 kstat_set_raw_ops(kstat_t *ksp,
     int (*headers)(char *buf, size_t size),
     int (*data)(char *buf, size_t size, void *data),
-    void *(*addr)(kstat_t *ksp, loff_t index))
+    void *(*addr)(kstat_t *ksp, off_t index))
 {
 	(void) ksp, (void) headers, (void) data, (void) addr;
 }
Index: lib/libzpool/include/sys/zfs_debug_os.h
===================================================================
--- lib/libzpool/include/sys/zfs_debug_os.h	(revision 62830)
+++ lib/libzpool/include/sys/zfs_debug_os.h	(working copy)
@@ -23,7 +23,8 @@
 #ifndef _SYS_ZFS_DEBUG_OS_H
 #define	_SYS_ZFS_DEBUG_OS_H
 
-#define	SET_ERROR(err) \
-	(__set_error(__FILE__, __func__, __LINE__, err), err)
+#define	SET_ERROR(err)	({ 					\
+	err;							\
+})
 
 #endif
Index: patches/dd-wrt-patches.patch
===================================================================
--- patches/dd-wrt-patches.patch	(revision 62833)
+++ patches/dd-wrt-patches.patch	(working copy)
@@ -461,3 +461,58 @@ Index: config/kernel-config-defined.m4
  dnl # Check CONFIG_ZLIB_INFLATE
  dnl #
  dnl # Verify the kernel has CONFIG_ZLIB_INFLATE support enabled.
+Index: include/os/linux/zfs/sys/zfs_debug_os.h
+===================================================================
+--- include/os/linux/zfs/sys/zfs_debug_os.h	(revision 62830)
++++ include/os/linux/zfs/sys/zfs_debug_os.h	(working copy)
+@@ -23,7 +23,8 @@
+ #ifndef _SYS_ZFS_DEBUG_OS_H
+ #define	_SYS_ZFS_DEBUG_OS_H
+ 
+-#define	 SET_ERROR(err) \
+-	(__set_error(__FILE__, __func__, __LINE__, err), err)
++#define	SET_ERROR(err)	({ 					\
++	err;							\
++})
+ 
+ #endif	/* _SYS_ZFS_DEBUG_OS_H */
+Index: lib/libspl/include/sys/kstat.h
+===================================================================
+--- lib/libspl/include/sys/kstat.h	(revision 62830)
++++ lib/libspl/include/sys/kstat.h	(working copy)
+@@ -86,6 +86,6 @@ extern void kstat_delete(kstat_t *);
+ extern void kstat_set_raw_ops(kstat_t *ksp,
+     int (*headers)(char *buf, size_t size),
+     int (*data)(char *buf, size_t size, void *data),
+-    void *(*addr)(kstat_t *ksp, loff_t index));
++    void *(*addr)(kstat_t *ksp, off_t index));
+ 
+ #endif	/* _SYS_KSTAT_H */
+Index: lib/libspl/kstat.c
+===================================================================
+--- lib/libspl/kstat.c	(revision 62830)
++++ lib/libspl/kstat.c	(working copy)
+@@ -58,7 +58,7 @@ void
+ kstat_set_raw_ops(kstat_t *ksp,
+     int (*headers)(char *buf, size_t size),
+     int (*data)(char *buf, size_t size, void *data),
+-    void *(*addr)(kstat_t *ksp, loff_t index))
++    void *(*addr)(kstat_t *ksp, off_t index))
+ {
+ 	(void) ksp, (void) headers, (void) data, (void) addr;
+ }
+Index: lib/libzpool/include/sys/zfs_debug_os.h
+===================================================================
+--- lib/libzpool/include/sys/zfs_debug_os.h	(revision 62830)
++++ lib/libzpool/include/sys/zfs_debug_os.h	(working copy)
+@@ -23,7 +23,8 @@
+ #ifndef _SYS_ZFS_DEBUG_OS_H
+ #define	_SYS_ZFS_DEBUG_OS_H
+ 
+-#define	SET_ERROR(err) \
+-	(__set_error(__FILE__, __func__, __LINE__, err), err)
++#define	SET_ERROR(err)	({ 					\
++	err;							\
++})
+ 
+ #endif
