Index: cmd/zed/agents/fmd_api.c
===================================================================
--- cmd/zed/agents/fmd_api.c	(revision 51025)
+++ cmd/zed/agents/fmd_api.c	(working copy)
@@ -36,7 +36,7 @@
 
 #include <sys/types.h>
 #include <sys/fm/protocol.h>
-#include <uuid/uuid.h>
+#include <uuid.h>
 #include <signal.h>
 #include <string.h>
 #include <time.h>
Index: cmd/zpool/os/linux/zpool_vdev_os.c
===================================================================
--- cmd/zpool/os/linux/zpool_vdev_os.c	(revision 51025)
+++ cmd/zpool/os/linux/zpool_vdev_os.c	(working copy)
@@ -82,8 +82,8 @@
 #include <sys/efi_partition.h>
 #include <sys/stat.h>
 #include <sys/mntent.h>
-#include <uuid/uuid.h>
-#include <blkid/blkid.h>
+#include <uuid.h>
+#include <blkid.h>
 
 typedef struct vdev_disk_db_entry
 {
Index: config/kernel-kuidgid.m4
===================================================================
--- config/kernel-kuidgid.m4	(revision 51025)
+++ config/kernel-kuidgid.m4	(working copy)
@@ -1,21 +1,28 @@
 dnl #
-dnl # 3.8 API change,
-dnl # User namespaces, use kuid_t in place of uid_t where available.
+dnl # User namespaces, use kuid_t in place of uid_t
+dnl # where available. Not strictly a user namespaces thing
+dnl # but it should prevent surprises
 dnl #
-AC_DEFUN([ZFS_AC_KERNEL_SRC_KUIDGID_T], [
-	ZFS_LINUX_TEST_SRC([kuidgid_t], [
+AC_DEFUN([ZFS_AC_KERNEL_KUIDGID_T], [
+	AC_MSG_CHECKING([whether kuid_t/kgid_t is available])
+	ZFS_LINUX_TRY_COMPILE([
 		#include <linux/uidgid.h>
 	], [
 		kuid_t userid __attribute__ ((unused)) = KUIDT_INIT(0);
 		kgid_t groupid __attribute__ ((unused)) = KGIDT_INIT(0);
-	])
-])
-
-AC_DEFUN([ZFS_AC_KERNEL_KUIDGID_T], [
-	AC_MSG_CHECKING([whether kuid_t/kgid_t is available])
-	ZFS_LINUX_TEST_RESULT([kuidgid_t], [
-		AC_MSG_RESULT(yes)
 	],[
-		ZFS_LINUX_TEST_ERROR([kuid_t/kgid_t])
+		ZFS_LINUX_TRY_COMPILE([
+			#include <linux/uidgid.h>
+		], [
+			kuid_t userid __attribute__ ((unused)) = 0;
+			kgid_t groupid __attribute__ ((unused)) = 0;
+		],[
+			AC_MSG_RESULT(yes; optional)
+		],[
+			AC_MSG_RESULT(yes; mandatory)
+			AC_DEFINE(HAVE_KUIDGID_T, 1, [kuid_t/kgid_t in use])
+		])
+	],[
+		AC_MSG_RESULT(no)
 	])
 ])
Index: config/kernel.m4
===================================================================
--- config/kernel.m4	(revision 51025)
+++ config/kernel.m4	(working copy)
@@ -121,7 +121,6 @@ AC_DEFUN([ZFS_AC_KERNEL_TEST_SRC], [
 	ZFS_AC_KERNEL_SRC_GENERIC_IO_ACCT
 	ZFS_AC_KERNEL_SRC_FPU
 	ZFS_AC_KERNEL_SRC_FMODE_T
-	ZFS_AC_KERNEL_SRC_KUIDGID_T
 	ZFS_AC_KERNEL_SRC_KUID_HELPERS
 	ZFS_AC_KERNEL_SRC_RENAME
 	ZFS_AC_KERNEL_SRC_CURRENT_TIME
Index: config/user-libblkid.m4
===================================================================
--- config/user-libblkid.m4	(revision 51025)
+++ config/user-libblkid.m4	(working copy)
@@ -3,7 +3,11 @@ dnl # Check for libblkid.  Basic support for detec
 dnl # has existing in blkid since 2008.
 dnl #
 AC_DEFUN([ZFS_AC_CONFIG_USER_LIBBLKID], [
-	ZFS_AC_FIND_SYSTEM_LIBRARY(LIBBLKID, [blkid], [blkid/blkid.h], [], [blkid], [], [], [
-		AC_MSG_FAILURE([
-		*** blkid.h missing, libblkid-devel package required])])
+	LIBBLKID=
+
+	AC_CHECK_HEADER([blkid.h], [], [AC_MSG_FAILURE([
+	*** blkid.h missing, libblkid-devel package required])])
+
+	AC_SUBST([LIBBLKID_LIBS], ["-lblkid"])
+	AC_DEFINE([HAVE_LIBBLKID], 1, [Define if you have libblkid])
 ])
Index: config/user-libuuid.m4
===================================================================
--- config/user-libuuid.m4	(revision 51025)
+++ config/user-libuuid.m4	(working copy)
@@ -2,7 +2,17 @@ dnl #
 dnl # Check for libuuid
 dnl #
 AC_DEFUN([ZFS_AC_CONFIG_USER_LIBUUID], [
-	ZFS_AC_FIND_SYSTEM_LIBRARY(LIBUUID, [uuid], [uuid/uuid.h], [], [uuid], [uuid_generate uuid_is_null], [], [
-	    AC_MSG_FAILURE([*** libuuid-devel package required])
-	])
+	LIBUUID=
+
+	AC_CHECK_HEADER([uuid.h], [], [AC_MSG_FAILURE([
+	*** uuid/uuid.h missing, libuuid-devel package required])])
+
+	AC_SEARCH_LIBS([uuid_generate], [uuid], [], [AC_MSG_FAILURE([
+	*** uuid_generate() missing, libuuid-devel package required])])
+
+	AC_SEARCH_LIBS([uuid_is_null], [uuid], [], [AC_MSG_FAILURE([
+	*** uuid_is_null() missing, libuuid-devel package required])])
+
+	AC_SUBST([LIBUUID], ["-luuid"])
+	AC_DEFINE([HAVE_LIBUUID], 1, [Define if you have libuuid])
 ])
Index: include/os/linux/kernel/linux/vfs_compat.h
===================================================================
--- include/os/linux/kernel/linux/vfs_compat.h	(revision 51025)
+++ include/os/linux/kernel/linux/vfs_compat.h	(working copy)
@@ -146,6 +146,11 @@ zpl_bdi_destroy(struct super_block *sb)
 #define	clear_inode(ip)		end_writeback(ip)
 #endif /* HAVE_EVICT_INODE && !HAVE_CLEAR_INODE */
 
+#ifndef SEEK_DATA
+#define SEEK_DATA 3
+#define SEEK_HOLE 4
+#endif
+
 #if defined(SEEK_HOLE) && defined(SEEK_DATA) && !defined(HAVE_LSEEK_EXECUTE)
 static inline loff_t
 lseek_execute(
Index: include/os/linux/spl/sys/cred.h
===================================================================
--- include/os/linux/spl/sys/cred.h	(revision 51025)
+++ include/os/linux/spl/sys/cred.h	(working copy)
@@ -42,6 +42,8 @@ extern struct task_struct init_task;
 #define	GROUP_AT(gi, i)	((gi)->gid[i])
 #endif
 
+#ifdef HAVE_KUIDGID_T
+
 #define	KUID_TO_SUID(x)		(__kuid_val(x))
 #define	KGID_TO_SGID(x)		(__kgid_val(x))
 #define	SUID_TO_KUID(x)		(KUIDT_INIT(x))
@@ -48,6 +50,16 @@ extern struct task_struct init_task;
 #define	SGID_TO_KGID(x)		(KGIDT_INIT(x))
 #define	KGIDP_TO_SGIDP(x)	(&(x)->val)
 
+#else /* HAVE_KUIDGID_T */
+
+#define	KUID_TO_SUID(x)		(x)
+#define	KGID_TO_SGID(x)		(x)
+#define	SUID_TO_KUID(x)		(x)
+#define	SGID_TO_KGID(x)		(x)
+#define	KGIDP_TO_SGIDP(x)	(x)
+
+#endif /* HAVE_KUIDGID_T */
+
 /* Check if the user ns is the initial one */
 static inline boolean_t
 zfs_is_init_userns(struct user_namespace *user_ns)
@@ -124,6 +136,7 @@ static inline gid_t zfs_vfsgid_to_gid(struct user_
 	return (__kgid_val(make_kgid(fs_userns, gid)));
 }
 
+
 extern void crhold(cred_t *cr);
 extern void crfree(cred_t *cr);
 extern uid_t crgetuid(const cred_t *cr);
Index: include/os/linux/spl/sys/mutex.h
===================================================================
--- include/os/linux/spl/sys/mutex.h	(revision 51025)
+++ include/os/linux/spl/sys/mutex.h	(working copy)
@@ -27,6 +27,7 @@
 #include <sys/types.h>
 #include <linux/sched.h>
 #include <linux/mutex.h>
+#include <linux/sched.h>
 #include <linux/lockdep.h>
 #include <linux/compiler_compat.h>
 
Index: include/os/linux/spl/sys/vnode.h
===================================================================
--- include/os/linux/spl/sys/vnode.h	(revision 51025)
+++ include/os/linux/spl/sys/vnode.h	(working copy)
@@ -52,6 +52,10 @@
 
 #define	F_FREESP	11 	/* Free file space */
 
+#ifndef SEEK_DATA
+#define SEEK_DATA 3
+#define SEEK_HOLE 4
+#endif
 
 #if defined(SEEK_HOLE) && defined(SEEK_DATA)
 #define	F_SEEK_DATA	SEEK_DATA
Index: lib/libefi/rdwr_efi.c
===================================================================
--- lib/libefi/rdwr_efi.c	(revision 51025)
+++ lib/libefi/rdwr_efi.c	(working copy)
@@ -30,7 +30,7 @@
 #include <errno.h>
 #include <string.h>
 #include <unistd.h>
-#include <uuid/uuid.h>
+#include <uuid.h>
 #include <zlib.h>
 #include <libintl.h>
 #include <sys/types.h>
Index: lib/libshare/os/linux/nfs.c
===================================================================
--- lib/libshare/os/linux/nfs.c	(revision 51025)
+++ lib/libshare/os/linux/nfs.c	(working copy)
@@ -29,6 +29,7 @@
 #include <dirent.h>
 #include <stdio.h>
 #include <string.h>
+#include <fcntl.h>
 #include <errno.h>
 #include <fcntl.h>
 #include <sys/file.h>
Index: lib/libuutil/Makefile.am
===================================================================
--- lib/libuutil/Makefile.am	(revision 51025)
+++ lib/libuutil/Makefile.am	(working copy)
@@ -17,7 +17,7 @@ libuutil_la_LIBADD = \
 
 libuutil_la_LIBADD += $(LTLIBINTL)
 
-libuutil_la_LDFLAGS = -pthread
+libuutil_la_LDFLAGS = -pthread -latomic
 
 if !ASAN_ENABLED
 libuutil_la_LDFLAGS += -Wl,-z,defs
Index: lib/libzfs_core/Makefile.am
===================================================================
--- lib/libzfs_core/Makefile.am	(revision 51025)
+++ lib/libzfs_core/Makefile.am	(working copy)
@@ -29,7 +29,7 @@ libzfs_core_la_LIBADD = \
 
 libzfs_core_la_LIBADD += $(LTLIBINTL)
 
-libzfs_core_la_LDFLAGS = -pthread
+libzfs_core_la_LDFLAGS = -pthread -latomic
 
 if !ASAN_ENABLED
 libzfs_core_la_LDFLAGS += -Wl,-z,defs
Index: lib/libzpool/Makefile.am
===================================================================
--- lib/libzpool/Makefile.am	(revision 51025)
+++ lib/libzpool/Makefile.am	(working copy)
@@ -195,7 +195,7 @@ libzpool_la_LIBADD = \
 
 libzpool_la_LIBADD += $(LIBCLOCK_GETTIME) $(ZLIB_LIBS) -ldl -lm
 
-libzpool_la_LDFLAGS = -pthread
+libzpool_la_LDFLAGS = -pthread -latomic
 
 if !ASAN_ENABLED
 libzpool_la_LDFLAGS += -Wl,-z,defs
Index: lib/libzutil/os/linux/zutil_import_os.c
===================================================================
--- lib/libzutil/os/linux/zutil_import_os.c	(revision 51025)
+++ lib/libzutil/os/linux/zutil_import_os.c	(working copy)
@@ -73,7 +73,7 @@
 #include <libudev.h>
 #include <sched.h>
 #endif
-#include <blkid/blkid.h>
+#include <blkid.h>
 
 #define	DEV_BYID_PATH	"/dev/disk/by-id/"
 
@@ -589,7 +589,7 @@ zfs_device_get_physical(struct udev_device *dev, c
 int
 zpool_label_disk_wait(const char *path, int timeout_ms)
 {
-#ifdef HAVE_LIBUDEV
+#if 0 //def HAVE_LIBUDEV
 	struct udev *udev;
 	struct udev_device *dev = NULL;
 	char nodepath[MAXPATHLEN];
Index: module/Kbuild.in
===================================================================
--- module/os/linux/zfs/zpl_file.c	(revision 51025)
+++ module/os/linux/zfs/zpl_file.c	(working copy)
@@ -552,6 +552,11 @@ zpl_direct_IO(int rw, struct kiocb *kiocb, struct
 
 #endif /* HAVE_VFS_RW_ITERATE */
 
+#ifndef SEEK_DATA
+#define SEEK_DATA 3
+#define SEEK_HOLE 4
+#endif
+
 static loff_t
 zpl_llseek(struct file *filp, loff_t offset, int whence)
 {
Index: module/zfs/zfs_vnops.c
===================================================================
--- module/zfs/zfs_vnops.c	(revision 51025)
+++ module/zfs/zfs_vnops.c	(working copy)
@@ -80,6 +80,10 @@ out:
 	return (error);
 }
 
+#ifndef SEEK_DATA
+#define SEEK_DATA 3
+#define SEEK_HOLE 4
+#endif
 
 #if defined(SEEK_HOLE) && defined(SEEK_DATA)
 /*
Index: tests/zfs-tests/cmd/mmap_seek.c
===================================================================
--- tests/zfs-tests/cmd/mmap_seek.c	(revision 51025)
+++ tests/zfs-tests/cmd/mmap_seek.c	(working copy)
@@ -35,6 +35,11 @@
 #include <linux/fs.h>
 #endif
 
+#ifndef SEEK_DATA
+#define SEEK_DATA 3
+#define SEEK_HOLE 4
+#endif
+
 static void
 seek_data(int fd, off_t offset, off_t expected)
 {
Index: tests/zfs-tests/cmd/readmmap.c
===================================================================
--- tests/zfs-tests/cmd/readmmap.c	(revision 51025)
+++ tests/zfs-tests/cmd/readmmap.c	(working copy)
@@ -45,6 +45,7 @@
 #include <errno.h>
 #include <sys/mman.h>
 #include <time.h>
+#include <sys/types.h>
 
 int
 main(int argc, char **argv)
Index: include/os/linux/spl/sys/cred.h
===================================================================
--- include/os/linux/spl/sys/cred.h	(revision 55236)
+++ include/os/linux/spl/sys/cred.h	(revision 55237)
@@ -42,6 +42,8 @@
 #define	GROUP_AT(gi, i)	((gi)->gid[i])
 #endif
 
+extern zidmap_t *zfs_get_init_idmap(void);
+
 #ifdef HAVE_KUIDGID_T
 
 #define	KUID_TO_SUID(x)		(__kuid_val(x))
@@ -50,7 +52,6 @@
 #define	SGID_TO_KGID(x)		(KGIDT_INIT(x))
 #define	KGIDP_TO_SGIDP(x)	(&(x)->val)
 
-extern zidmap_t *zfs_get_init_idmap(void);
 
 #else /* HAVE_KUIDGID_T */
 
Index: config/kernel.m4
===================================================================
--- config/kernel.m4	(revision 55848)
+++ config/kernel.m4	(working copy)
@@ -712,7 +712,7 @@ AC_DEFUN([ZFS_LINUX_COMPILE], [
 	    KBUILD_MODPOST_NOFINAL="$5" KBUILD_MODPOST_WARN="$6"
 	    make modules -k -j$TEST_JOBS ${KERNEL_CC:+CC=$KERNEL_CC}
 	    ${KERNEL_LD:+LD=$KERNEL_LD} ${KERNEL_LLVM:+LLVM=$KERNEL_LLVM}
-	    CONFIG_MODULES=y CFLAGS_MODULE=-DCONFIG_MODULES
+	    CONFIG_MODULES=y CFLAGS_MODULE="-DCONFIG_MODULES -fno-lto"
 	    -C $LINUX_OBJ $ARCH_UM M=$PWD/$1 >$1/build.log 2>&1])
 	AS_IF([AC_TRY_COMMAND([$2])], [$3], [$4])
 ])
Index: module/Kbuild.in
===================================================================
--- module/Kbuild.in	(revision 55862)
+++ module/Kbuild.in	(working copy)
@@ -2,7 +2,7 @@
 # first.  This ensures its module initialization function is run before
 # any of the other module initialization functions which depend on it.
 
-ZFS_MODULE_CFLAGS += -std=gnu99 -Wno-declaration-after-statement
+ZFS_MODULE_CFLAGS += -std=gnu99 -Wno-declaration-after-statement -fno-lto
 ZFS_MODULE_CFLAGS += -Wmissing-prototypes
 ZFS_MODULE_CFLAGS += @KERNEL_DEBUG_CFLAGS@  @NO_FORMAT_ZERO_LENGTH@
 
Index: lib/libspl/include/sys/backtrace.h
===================================================================
--- lib/libspl/include/sys/backtrace.h	(revision 56411)
+++ lib/libspl/include/sys/backtrace.h	(working copy)
@@ -27,6 +27,10 @@
 #ifndef _LIBSPL_SYS_BACKTRACE_H
 #define	_LIBSPL_SYS_BACKTRACE_H
 
+#ifndef __maybe_unused
+#define	__maybe_unused __attribute__((unused))
+#endif
+
 void libspl_backtrace(int fd);
 
 #endif
