Index: module/Kbuild.in
===================================================================
--- module/Kbuild.in	(revision 52656)
+++ module/Kbuild.in	(working copy)
@@ -43,9 +43,9 @@ asflags-y := $(ZFS_MODULE_CFLAGS) $(ZFS_MODULE_CPP
 ccflags-y := $(ZFS_MODULE_CFLAGS) $(ZFS_MODULE_CPPFLAGS)
 
 ifeq ($(CONFIG_ARM64),y)
-CFLAGS_REMOVE_zcommon/zfs_fletcher_aarch64_neon.o += -mgeneral-regs-only
-CFLAGS_REMOVE_zfs/vdev_raidz_math_aarch64_neon.o += -mgeneral-regs-only
-CFLAGS_REMOVE_zfs/vdev_raidz_math_aarch64_neonx2.o += -mgeneral-regs-only
+CFLAGS_REMOVE_zfs_fletcher_aarch64_neon.o = -mgeneral-regs-only
+CFLAGS_REMOVE_vdev_raidz_math_aarch64_neon.o = -mgeneral-regs-only
+CFLAGS_REMOVE_vdev_raidz_math_aarch64_neonx2.o = -mgeneral-regs-only
 endif
 
 # Suppress unused-value warnings in sparc64 architecture headers
