Index: include/os/linux/spl/sys/debug.h
===================================================================
--- include/os/linux/spl/sys/debug.h	(revision 49354)
+++ include/os/linux/spl/sys/debug.h	(working copy)
@@ -66,31 +66,31 @@ void spl_dumpstack(void);
 
 #define	VERIFY3B(LEFT, OP, RIGHT)					\
 do {									\
-	const boolean_t __left = (boolean_t)(LEFT);			\
-	const boolean_t __right = (boolean_t)(RIGHT);			\
+	const boolean_t __left __attribute__((unused)) = (boolean_t)(LEFT);			\
+	const boolean_t __right __attribute__((unused)) = (boolean_t)(RIGHT);			\
 } while (0)
 
 #define	VERIFY3S(LEFT, OP, RIGHT)					\
 do {									\
-	const int64_t __left = (int64_t)(LEFT);				\
-	const int64_t __right = (int64_t)(RIGHT);			\
+	const int64_t __left __attribute__((unused)) = (int64_t)(LEFT);				\
+	const int64_t __right __attribute__((unused)) = (int64_t)(RIGHT);			\
 } while (0)
 
 #define	VERIFY3U(LEFT, OP, RIGHT)					\
 do {									\
-	const uint64_t __left = (uint64_t)(LEFT);			\
-	const uint64_t __right = (uint64_t)(RIGHT);			\
+	const uint64_t __left __attribute__((unused)) = (uint64_t)(LEFT);			\
+	const uint64_t __right __attribute__((unused)) = (uint64_t)(RIGHT);			\
 } while (0)
 
 #define	VERIFY3P(LEFT, OP, RIGHT)					\
 do {									\
-	const uintptr_t __left = (uintptr_t)(LEFT);			\
-	const uintptr_t __right = (uintptr_t)(RIGHT);			\
+	const uintptr_t __left __attribute__((unused)) = (uintptr_t)(LEFT);			\
+	const uintptr_t __right __attribute__((unused)) = (uintptr_t)(RIGHT);			\
 } while (0)
 
 #define	VERIFY0(LEFT)							\
 do {									\
-	const uint64_t __left = (uint64_t)(LEFT);			\
+	const uint64_t __left __attribute__((unused)) = (uint64_t)(LEFT);			\
 } while (0)
 #else
 #define	VERIFY(cond)							\
Index: lib/libspl/include/assert.h
===================================================================
--- lib/libspl/include/assert.h	(revision 49354)
+++ lib/libspl/include/assert.h	(working copy)
@@ -57,31 +57,31 @@ libspl_assert(const char *buf, const char *file, c
 
 #define	VERIFY3B(LEFT, OP, RIGHT)					\
 do {									\
-	const boolean_t __left = (boolean_t)(LEFT);			\
-	const boolean_t __right = (boolean_t)(RIGHT);			\
+	const boolean_t __left __attribute__((unused)) = (boolean_t)(LEFT);			\
+	const boolean_t __right __attribute__((unused)) = (boolean_t)(RIGHT);			\
 } while (0)
 
 #define	VERIFY3S(LEFT, OP, RIGHT)					\
 do {									\
-	const int64_t __left = (int64_t)(LEFT);				\
-	const int64_t __right = (int64_t)(RIGHT);			\
+	const int64_t __left __attribute__((unused)) = (int64_t)(LEFT);				\
+	const int64_t __right __attribute__((unused)) = (int64_t)(RIGHT);			\
 } while (0)
 
 #define	VERIFY3U(LEFT, OP, RIGHT)					\
 do {									\
-	const uint64_t __left = (uint64_t)(LEFT);			\
-	const uint64_t __right = (uint64_t)(RIGHT);			\
+	const uint64_t __left __attribute__((unused)) = (uint64_t)(LEFT);			\
+	const uint64_t __right __attribute__((unused)) = (uint64_t)(RIGHT);			\
 } while (0)
 
 #define	VERIFY3P(LEFT, OP, RIGHT)					\
 do {									\
-	const uintptr_t __left = (uintptr_t)(LEFT);			\
-	const uintptr_t __right = (uintptr_t)(RIGHT);			\
+	const uintptr_t __left __attribute__((unused)) = (uintptr_t)(LEFT);			\
+	const uintptr_t __right __attribute__((unused)) = (uintptr_t)(RIGHT);			\
 } while (0)
 
 #define	VERIFY0(LEFT)							\
 do {									\
-	const uint64_t __left = (uint64_t)(LEFT);			\
+	const uint64_t __left __attribute__((unused)) = (uint64_t)(LEFT);			\
 } while (0)
 #else
 #define	VERIFY(cond)							\
