Index: include/os/linux/spl/sys/debug.h
===================================================================
--- include/os/linux/spl/sys/debug.h	(revision 49240)
+++ include/os/linux/spl/sys/debug.h	(working copy)
@@ -61,6 +61,15 @@ void spl_dumpstack(void);
 #define	PANIC(fmt, a...)						\
 	spl_panic(__FILE__, __FUNCTION__, __LINE__, fmt, ## a)
 
+#ifdef NDEBUG
+#define	VERIFY(cond) ((void) sizeof ((uintptr_t)(cond)))
+#define	verify(cond) ((void) sizeof ((uintptr_t)(cond)))
+#define	VERIFY3B(LEFT, OP, RIGHT) ((void) sizeof ((uintptr_t)(LEFT)), (void) sizeof ((uintptr_t)(RIGHT)))
+#define	VERIFY3S(LEFT, OP, RIGHT) ((void) sizeof ((uintptr_t)(LEFT)), (void) sizeof ((uintptr_t)(RIGHT)))
+#define	VERIFY3U(LEFT, OP, RIGHT) ((void) sizeof ((uintptr_t)(LEFT)), (void) sizeof ((uintptr_t)(RIGHT)))
+#define	VERIFY3P(LEFT, OP, RIGHT) ((void) sizeof ((uintptr_t)(LEFT)), (void) sizeof ((uintptr_t)(RIGHT)))
+#define	VERIFY0(LEFT) ((void) sizeof ((uintptr_t)(LEFT)))
+#else
 #define	VERIFY(cond)							\
 	(void) (unlikely(!(cond)) &&					\
 	    spl_panic(__FILE__, __FUNCTION__, __LINE__,			\
@@ -119,7 +128,7 @@ void spl_dumpstack(void);
 		    "failed (0 == %lld)\n",				\
 		    (long long) (_verify3_right));			\
 	} while (0)
-
+#endif
 /*
  * Debugging disabled (--disable-debug)
  */
Index: lib/libspl/include/assert.h
===================================================================
--- lib/libspl/include/assert.h	(revision 49240)
+++ lib/libspl/include/assert.h	(working copy)
@@ -51,7 +51,16 @@ libspl_assert(const char *buf, const char *file, c
 #ifdef verify
 #undef verify
 #endif
+#ifdef NDEBUG
+#define	VERIFY(cond) ((void) sizeof ((uintptr_t)(cond)))
+#define	verify(cond) ((void) sizeof ((uintptr_t)(cond)))
+#define	VERIFY3B(LEFT, OP, RIGHT) ((void) sizeof ((uintptr_t)(LEFT)), (void) sizeof ((uintptr_t)(RIGHT)))
+#define	VERIFY3S(LEFT, OP, RIGHT) ((void) sizeof ((uintptr_t)(LEFT)), (void) sizeof ((uintptr_t)(RIGHT)))
+#define	VERIFY3U(LEFT, OP, RIGHT) ((void) sizeof ((uintptr_t)(LEFT)), (void) sizeof ((uintptr_t)(RIGHT)))
+#define	VERIFY3P(LEFT, OP, RIGHT) ((void) sizeof ((uintptr_t)(LEFT)), (void) sizeof ((uintptr_t)(RIGHT)))
+#define	VERIFY0(LEFT) ((void) sizeof ((uintptr_t)(LEFT)))
 
+#else
 #define	VERIFY(cond)							\
 	(void) ((!(cond)) &&						\
 	    libspl_assert(#cond, __FILE__, __FUNCTION__, __LINE__))
@@ -107,7 +116,7 @@ do {									\
 		    "%s == 0 (0x%llx == 0)", #LEFT,			\
 		    (u_longlong_t)__left);				\
 } while (0)
-
+#endif
 #ifdef assert
 #undef assert
 #endif
