Index: etc/systemd/system-generators/zfs-mount-generator.c
===================================================================
--- etc/systemd/system-generators/zfs-mount-generator.c	(revision 46971)
+++ etc/systemd/system-generators/zfs-mount-generator.c	(working copy)
@@ -206,6 +206,8 @@
 	return (fdopen(fd, stream_mode));
 }
 
+#define	my_strndupa(x,y)	strncpy(alloca(y),x,y)
+
 static int
 line_worker(char *line, const char *cachefile)
 {
@@ -235,7 +237,7 @@
 
 	const char *pool = dataset;
 	if ((toktmp = strchr(pool, '/')) != NULL)
-		pool = strndupa(pool, toktmp - pool);
+		pool = my_strndupa(pool, toktmp - pool);
 
 	if (p_nbmand == NULL) {
 		fprintf(stderr, PROGNAME "[%d]: %s: not enough tokens!\n",
