Index: lib/libzutil/os/linux/zutil_import_os.c
===================================================================
--- lib/libzutil/os/linux/zutil_import_os.c	(revision 45659)
+++ lib/libzutil/os/linux/zutil_import_os.c	(working copy)
@@ -606,7 +606,7 @@
 int
 zpool_label_disk_wait(const char *path, int timeout_ms)
 {
-#ifdef HAVE_LIBUDEV
+#if 0 //def HAVE_LIBUDEV
 	struct udev *udev;
 	struct udev_device *dev = NULL;
 	char nodepath[MAXPATHLEN];
