Index: ndpi_protocol_ids.h
===================================================================
--- ndpi_protocol_ids.h	(revision 54643)
+++ ndpi_protocol_ids.h	(working copy)
@@ -100,7 +100,7 @@
   NDPI_PROTOCOL_AMONG_US              = 69,
   NDPI_PROTOCOL_YAHOO                 = 70,
   NDPI_PROTOCOL_DISNEYPLUS            = 71,
-  NDPI_PROTOCOL_GOOGLE_PLUS           = 72,
+  NDPI_PROTOCOL_HART_IP               = 72,
   NDPI_PROTOCOL_IP_VRRP               = 73,
   NDPI_PROTOCOL_STEAM                 = 74,
   NDPI_PROTOCOL_HALFLIFE2             = 75,
@@ -365,7 +365,6 @@
   NDPI_PROTOCOL_BACNET                = 334,
   NDPI_PROTOCOL_OICQ                  = 335,
   NDPI_PROTOCOL_HOTS                  = 336, /* Heroes of the Storm */
-
   NDPI_PROTOCOL_FACEBOOK_REEL_STORY   = 337,
   NDPI_PROTOCOL_SRTP                  = 338,
   NDPI_PROTOCOL_OPERA_VPN             = 339,
@@ -380,6 +379,28 @@
   NDPI_PROTOCOL_MULLVAD               = 348,
   NDPI_PROTOCOL_HTTP2                 = 349,
   NDPI_PROTOCOL_HAPROXY               = 350,
+  NDPI_PROTOCOL_RMCP                  = 351,
+  NDPI_PROTOCOL_CAN                   = 352,
+  NDPI_PROTOCOL_PROTOBUF              = 353,
+  NDPI_PROTOCOL_ETHEREUM              = 354, 
+  NDPI_PROTOCOL_TELEGRAM_VOIP         = 355,
+  NDPI_PROTOCOL_SINA_WEIBO            = 356,
+  NDPI_PROTOCOL_TESLA_SERVICES        = 357,
+  NDPI_PROTOCOL_PTPV2                 = 358,
+  NDPI_PROTOCOL_RTPS                  = 359,
+  NDPI_PROTOCOL_OPC_UA                = 360,
+  NDPI_PROTOCOL_S7COMM_PLUS           = 361,
+  NDPI_PROTOCOL_FINS                  = 362,
+  NDPI_PROTOCOL_ETHERSIO              = 363,
+  NDPI_PROTOCOL_UMAS                  = 364,
+  NDPI_PROTOCOL_BECKHOFF_ADS          = 365,
+  NDPI_PROTOCOL_ISO9506_1_MMS         = 366,
+  NDPI_PROTOCOL_IEEE_C37118           = 367,
+  NDPI_PROTOCOL_ETHERSBUS             = 368,
+  NDPI_PROTOCOL_MONERO                = 369,
+  NDPI_PROTOCOL_DCERPC                = 370,
+  NDPI_PROTOCOL_PROFINET_IO           = 371,
+  NDPI_PROTOCOL_HISLIP                = 372,
 
 #ifdef CUSTOM_NDPI_PROTOCOLS
 #include "../../../nDPI-custom/custom_ndpi_protocol_ids.h"
