diff -puNrb linux-2.6.22.14/drivers/net/ucc_geth.h linux/drivers/net/ucc_geth.h
--- linux-2.6.22.14/drivers/net/ucc_geth.h	2008-01-22 14:03:03.000000000 +0200
+++ linux/drivers/net/ucc_geth.h	2008-01-15 11:04:02.000000000 +0200
@@ -900,8 +900,8 @@ struct ucc_geth_hardware_statistics {
 #define UCC_GETH_SIZE_OF_BD                     QE_SIZEOF_BD
 
 /* Driver definitions */
-#define TX_BD_RING_LEN                          0x10
-#define RX_BD_RING_LEN                          0x10
+#define TX_BD_RING_LEN                          0x80
+#define RX_BD_RING_LEN                          0x80
 #define UCC_GETH_DEV_WEIGHT                     TX_BD_RING_LEN
 
 #define TX_RING_MOD_MASK(size)                  (size-1)
@@ -1242,6 +1242,7 @@ struct ucc_geth_private {
 	int oldspeed;
 	int oldduplex;
 	int oldlink;
+	unsigned long last_tx;
 };
 
 #endif				/* __UCC_GETH_H__ */
