Index: samba-3.0.24/source/lib/popt_common.c
===================================================================
--- samba-3.0.24.orig/source/lib/popt_common.c	2008-08-11 11:25:26.000000000 +0200
+++ samba-3.0.24/source/lib/popt_common.c	2008-08-11 11:26:03.000000000 +0200
@@ -54,7 +54,7 @@
 	else 
 		pname++;
 
-	pstr_sprintf(logfile, "%s/log.%s", arg, pname);
+	pstr_sprintf(logfile, "%s/log/%s.log", arg, pname);
 	lp_set_logfile(logfile);
 }
 
