--- services.c	2005-05-17 21:18:27.000000000 +0000
+++ ../../../../DD-WRT1/src/router/rc/services.c	2005-06-04 16:38:05.000000000 +0000
@@ -1977,6 +1977,10 @@
         symlink("/sbin/rc", "/tmp/ppp/ip-down");
         symlink("/sbin/rc", "/tmp/ppp/set-pppoepid"); // tallest 1219
         unlink("/tmp/ppp/log");
+		
+	//Clean rpppoe client files - Added by ice-man (Wed Jun 1)
+	unlink("/tmp/ppp/options.pppoe");
+	unlink("/tmp/ppp/connect-errors");
                                                                                                                              
         _eval(pppoe_argv, NULL, 0, &pid);
                                                                                                                              
