diff -puNrb linux-2.6.22.14/arch/mips/mm/c-r4k.c linux/arch/mips/mm/c-r4k.c
--- linux-2.6.22.14/arch/mips/mm/c-r4k.c	2008-01-22 14:00:33.000000000 +0200
+++ linux/arch/mips/mm/c-r4k.c	2008-01-15 11:01:18.000000000 +0200
@@ -890,7 +890,9 @@ static void __init probe_pcache(void)
 		              c->dcache.linesz;
 		c->dcache.waybit = __ffs(dcache_size/c->dcache.ways);
 
+#ifdef CONFIG_CPU_HAS_PREFETCH
 		c->options |= MIPS_CPU_PREFETCH;
+#endif
 		break;
 	}
 
