diff -puNrb linux-2.6.27.39/include/linux/mtd/nand.h linux/include/linux/mtd/nand.h
--- linux-2.6.27.39/include/linux/mtd/nand.h	2009-11-20 09:52:56.000000000 +0200
+++ linux/include/linux/mtd/nand.h	2009-11-20 09:55:50.000000000 +0200
@@ -404,6 +404,9 @@ struct nand_chip {
 	int		subpagesize;
 	uint8_t		cellinfo;
 	int		badblockpos;
+#ifdef MIPSEL
+	unsigned	backup_offset;
+#endif
 
 	nand_state_t	state;
 
@@ -426,6 +429,14 @@ struct nand_chip {
 	void		*priv;
 };
 
+#define BACKUP_4xFF_OFFSET	36
+#define ALWAYS_4x00_OFFSET	32
+#define ECC_ID_OFFSET		01
+/* ECC_ID_OFFSET contains:
+ *	0x00 => mlc ecc (>= 4bit/512 bytes),
+ *	0xff => slc ecc (>= 1bit/256 bytes)
+ */
+
 /*
  * NAND Flash Manufacturer ID Codes
  */
