diff -Naurp -x mach-bcm963xx linux-2.6.16.7-generic-patched/include/asm-mips/module.h linux-2.6.16.7-patched/include/asm-mips/module.h
--- linux-2.6.16.7-generic-patched/include/asm-mips/module.h	2006-04-17 23:53:25.000000000 +0200
+++ linux-2.6.16.7-patched/include/asm-mips/module.h	2006-07-05 15:21:58.000000000 +0200
@@ -113,6 +113,12 @@ search_module_dbetables(unsigned long ad
 #define MODULE_PROC_FAMILY "RM9000 "
 #elif defined CONFIG_CPU_SB1
 #define MODULE_PROC_FAMILY "SB1 "
+#elif defined CONFIG_CPU_BCM6338
+#define MODULE_PROC_FAMILY "BCM6338 "
+#elif defined CONFIG_CPU_BCM6345
+#define MODULE_PROC_FAMILY "BCM6345 "
+#elif defined CONFIG_CPU_BCM6348
+#define MODULE_PROC_FAMILY "BCM6348 "
 #else
 #error MODULE_PROC_FAMILY undefined for your processor configuration
 #endif
