--- a/drivers/net/phy/ar8216.c
+++ b/drivers/net/phy/ar8216.c
@@ -488,6 +488,12 @@ ar8216_reset_switch(struct switch_dev *d
 	}
 	/* XXX: undocumented magic from atheros, required! */
 	priv->write(priv, 0x38, 0xc000050e);
+
+	/* set mtu */
+	ar8216_rmw(priv, AR8216_REG_GLOBAL_CTRL,
+		AR8216_GCTRL_MTU,
+		1500 + 4 /* vlan */ + 2 /* header */);
+
 	return ar8216_hw_apply(dev);
 }
 
--- a/drivers/net/phy/ar8216.h
+++ b/drivers/net/phy/ar8216.h
@@ -26,6 +26,9 @@
 #define AR8216_REG_CTRL			0x0000
 #define   AR8216_CTRL_RESET		BIT(31)
 
+#define AR8216_REG_GLOBAL_CTRL		0x0030
+#define   AR8216_GCTRL_MTU		BITS(0, 10)
+
 #define AR8216_REG_VTU			0x0040
 #define   AR8216_VTU_OP			BITS(0, 3)
 #define   AR8216_VTU_OP_NOOP		0x0
