diff -Naurp -x brcm-boards -x pci linux-2.6.16.7-generic-patched/arch/mips/mm/c-r4k.c linux-2.6.16.7-patched/arch/mips/mm/c-r4k.c
--- linux-2.6.16.7-generic-patched/arch/mips/mm/c-r4k.c	2006-04-17 23:53:25.000000000 +0200
+++ linux-2.6.16.7-patched/arch/mips/mm/c-r4k.c	2006-07-05 15:21:58.000000000 +0200
@@ -889,6 +889,12 @@ static void __init probe_pcache(void)
 		if (!(config & MIPS_CONF_M))
 			panic("Don't know how to probe P-caches on this cpu.");
 
+        	if (c->cputype == CPU_BCM6338 || c->cputype == CPU_BCM6345 || c->cputype == CPU_BCM6348){
+			printk("brcm mips: enabling icache and dcache...\n");
+        	        /* Enable caches */
+        	        write_c0_diag(read_c0_diag() | 0xC0000000);
+        	}
+
 		/*
 		 * So we seem to be a MIPS32 or MIPS64 CPU
 		 * So let's probe the I-cache ...
