--- linux-2.6.35/arch/arm/include/asm/spinlock_types_sop.h	1970-01-01 01:00:00.000000000 +0100
+++ linux-2.6.35.12/arch/arm/include/asm/spinlock_types_sop.h	2011-05-10 09:35:03.000000000 +0200
@@ -0,0 +1,20 @@
+#ifndef __ASM_SPINLOCK_TYPES_H
+#define __ASM_SPINLOCK_TYPES_H
+
+#ifndef __LINUX_SPINLOCK_TYPES_H
+# error "please don't include this file directly"
+#endif
+
+typedef struct {
+	volatile unsigned int lock;
+} raw_spinlock_smp_t;
+
+#define __RAW_SPIN_LOCK_UNLOCKED_SMP	{ 0 }
+
+typedef struct {
+	volatile unsigned int lock;
+} raw_rwlock_smp_t;
+
+#define __RAW_RW_LOCK_UNLOCKED_SMP	{ 0 }
+
+#endif
