--- nocatsplash/S60nocat	1970-01-01 01:00:00.000000000 +0100
+++ nocatsplash/S61nocat	2005-05-23 10:04:04.000000000 +0100
@@ -0,0 +1,31 @@
+#!/bin/sh
+#
+# start/stop splashd super server.
+case "$1" in
+    start)
+	echo -n "Starting NoCat splash server:"
+	echo -n " splashd" ; 
+	/usr/sbin/splashd -D &
+	echo "."
+	;;
+    stop)
+	echo -n "Stopping NoCat splash server:"
+	echo -n " splashd" ; 
+	killall splashd
+	echo "."
+	;;
+    restart)
+	echo -n "Restarting NoCat splash server:"
+	echo -n " splashd"
+	$0 stop
+	$0 start
+	echo "."
+	;;
+    *)
+	echo "Usage: $0 {start|stop|restart}"
+	exit 1
+	;;
+esac
+
+exit 0
+
