diff -puNrb linux-2.6.27.39/include/linux/mtd/nand_ecc_mlc.h linux/include/linux/mtd/nand_ecc_mlc.h
--- linux-2.6.27.39/include/linux/mtd/nand_ecc_mlc.h	1970-01-01 03:00:00.000000000 +0300
+++ linux/include/linux/mtd/nand_ecc_mlc.h	2009-11-20 09:55:50.000000000 +0200
@@ -0,0 +1,7 @@
+#ifndef __MTD_NAND_ECC_MLC_H__
+#define __MTD_NAND_ECC_MLC_H__
+
+int nand_calculate_ecc_mlc(const u_char *dat, u_char *ecc_code);
+int nand_correct_data_mlc(u_char *dat, u_char *read_ecc);
+
+#endif /* __MTD_NAND_ECC_H__ */
