diff -urN SDK_3_1_0_0/RT288x_SDK/source/linux-2.6.21.x/drivers/net/raeth/Makefile SDK_3_2_0_0/RT288x_SDK/source/linux-2.6.21.x//drivers/net/raeth/Makefile
--- SDK_3_1_0_0/RT288x_SDK/source/linux-2.6.21.x/drivers/net/raeth/Makefile	2008-08-06 05:27:47.000000000 +0200
+++ SDK_3_2_0_0/RT288x_SDK/source/linux-2.6.21.x//drivers/net/raeth/Makefile	2008-10-24 08:20:14.000000000 +0200
@@ -1,5 +1,5 @@
 obj-$(CONFIG_RAETH) += raeth.o
-raeth-objs := raether.o ra_mac.o mcast.o
+raeth-objs := raether.o ra_mac.o mcast.o mii_mgr.o
 
 EXTRA_CFLAGS   += 
 
@@ -7,6 +7,7 @@
 raeth-objs += ra_qos.o
 endif
 
+ifeq ($(CONFIG_ETHTOOL),y)
 ifeq ($(CONFIG_RAETH_ROUTER),y)
 raeth-objs += ra_ethtool.o
 else
@@ -14,6 +15,7 @@
 raeth-objs += ra_ethtool.o
 endif
 endif
+endif
 
 ifeq ($(CONFIG_RAETH_NETLINK),y)
 raeth-objs += ra_netlink.o
@@ -38,3 +40,7 @@
 ifeq ($(CONFIG_RAETH_JUMBOFRAME),y)
 EXTRA_CFLAGS += -DCONFIG_RAETH_JUMBOFRAME
 endif
+
+ifeq ($(CONFIG_RT2860V2_AP_MEMORY_OPTIMIZATION),y)
+EXTRA_CFLAGS   += -DMEMORY_OPTIMIZATION
+endif
