diff -urN SDK_3_1_0_0/RT288x_SDK/source/linux-2.6.21.x/drivers/net/raeth/ra2882ethreg.h SDK_3_2_0_0/RT288x_SDK/source/linux-2.6.21.x//drivers/net/raeth/ra2882ethreg.h
--- SDK_3_1_0_0/RT288x_SDK/source/linux-2.6.21.x/drivers/net/raeth/ra2882ethreg.h	2008-08-06 05:27:47.000000000 +0200
+++ SDK_3_2_0_0/RT288x_SDK/source/linux-2.6.21.x//drivers/net/raeth/ra2882ethreg.h	2008-10-24 08:20:14.000000000 +0200
@@ -318,6 +318,9 @@
 #define RT2880_RX_DMA_EN         RT2880_BIT(2)
 #define RT2880_TX_DMA_EN         RT2880_BIT(0)
 
+#define PDMA_BT_SIZE_4DWORDS     (0<<4)
+#define PDMA_BT_SIZE_8DWORDS     (1<<4)
+#define PDMA_BT_SIZE_16DWORDS    (2<<4)
 
 /* Register bits.
  */
@@ -586,7 +589,9 @@
     struct net_device *PseudoDev;
     unsigned int isPseudo;
 #endif
+#if defined (CONFIG_ETHTOOL) && ( defined (CONFIG_RAETH_ROUTER) || defined (CONFIG_RT_3052_ESW) )
 	struct mii_if_info	mii_info;
+#endif
 } END_DEVICE, *pEND_DEVICE;
 
 #define RAETH_VERSION	"v1.60"
