diff -urN SDK_3_1_0_0/RT288x_SDK/source/linux-2.6.21.x/drivers/net/raeth/ra_ioctl.h SDK_3_2_0_0/RT288x_SDK/source/linux-2.6.21.x//drivers/net/raeth/ra_ioctl.h
--- SDK_3_1_0_0/RT288x_SDK/source/linux-2.6.21.x/drivers/net/raeth/ra_ioctl.h	2008-08-06 05:27:47.000000000 +0200
+++ SDK_3_2_0_0/RT288x_SDK/source/linux-2.6.21.x//drivers/net/raeth/ra_ioctl.h	2008-10-24 08:20:14.000000000 +0200
@@ -4,6 +4,8 @@
 /* ioctl commands */
 #define RAETH_ESW_REG_READ		0x89F1
 #define RAETH_ESW_REG_WRITE		0x89F2
+#define RAETH_MII_READ			0x89F3
+#define RAETH_MII_WRITE			0x89F4
 
 
 /* rt3052 embedded ethernet switch registers */
@@ -25,5 +27,12 @@
 	unsigned int val;
 } esw_reg;
 
+typedef struct ralink_mii_ioctl_data {
+	__u32	phy_id;
+	__u32	reg_num;
+	__u32	val_in;
+	__u32	val_out;
+} ra_mii_ioctl_data;
+
 
 #endif
