diff -urN SDK_3_1_0_0/RT288x_SDK/source/linux-2.6.21.x/drivers/net/raeth/ra_mac.h SDK_3_2_0_0/RT288x_SDK/source/linux-2.6.21.x//drivers/net/raeth/ra_mac.h
--- SDK_3_1_0_0/RT288x_SDK/source/linux-2.6.21.x/drivers/net/raeth/ra_mac.h	2008-08-06 05:27:47.000000000 +0200
+++ SDK_3_2_0_0/RT288x_SDK/source/linux-2.6.21.x//drivers/net/raeth/ra_mac.h	2008-10-24 08:20:14.000000000 +0200
@@ -21,10 +21,10 @@
 unsigned long readreg(unsigned long addr);
 void writereg(unsigned long value, unsigned long addr);
 
-int phyregread(unsigned int phy_register, unsigned int* value);
-int phyregwrite(unsigned int phy_register, unsigned int value);
-void rt2880_mdio_cfg(unsigned int address, unsigned int offset, unsigned int value);
-void phySetup(void);
+void enable_auto_negotiate(void);
+int wait_linkup(void);
+int wait_an_completed(void);
+
 void rt2880_gmac_hard_reset(void);
 
 extern u32 mii_mgr_read( unsigned int , unsigned int, unsigned int *);
