diff -urN SDK_3_1_0_0/RT288x_SDK/source/linux-2.6.21.x/drivers/net/raeth/ra_qos.c SDK_3_2_0_0/RT288x_SDK/source/linux-2.6.21.x//drivers/net/raeth/ra_qos.c
--- SDK_3_1_0_0/RT288x_SDK/source/linux-2.6.21.x/drivers/net/raeth/ra_qos.c	2008-08-06 05:27:47.000000000 +0200
+++ SDK_3_2_0_0/RT288x_SDK/source/linux-2.6.21.x//drivers/net/raeth/ra_qos.c	2008-10-24 08:20:14.000000000 +0200
@@ -302,9 +302,9 @@
    DSCP | AC | WMM_AC (Access Category)
    ------+----+--------
    00-07|  1 |  BE
-   08-15|  1 |  BE
+   24-31|  1 |  BE
+   08-15|  0 |  BG
    16-23|  0 |  BG
-   24-31|  0 |  BG
    32-39|  2 |  VI
    40-47|  2 |  VI
    48-55|  3 |  VO
@@ -324,7 +324,7 @@
 
     /* Bridge:: {BG,BE,VI,VO} */
     /* GateWay:: WAN: {BG,BE,VI,VO}, LAN: {BG,BE,VI,VO} */
-#if defined (CONFIG_RALINK_RT2883) 
+#if defined (CONFIG_RALINK_RT2883) && defined (CONFIG_RAETH_GMAC2)
     /* 
      * 1) Bridge: 
      *    1.1) GMAC1 ONLY:
@@ -343,7 +343,7 @@
     static unsigned char AcToRing_BridgeMap[4] = {2, 2, 3, 3}; 
     static unsigned char AcToRing_GE1Map[2][4] = {{3, 3, 3, 3},{2, 2, 2, 2}}; 
     static unsigned char AcToRing_GE2Map[4] = {0, 0, 1, 1};
-#elif defined (CONFIG_RALINK_RT3052) 
+#elif defined (CONFIG_RALINK_RT3052) || (defined (CONFIG_RALINK_RT2883) && !defined(CONFIG_RAETH_GMAC2))
     /* 
      * 1) Bridge: VO->Ring3, VI->Ring2, BG->Ring1, BE->Ring0 
      * 2) GateWay:
@@ -397,16 +397,16 @@
 	    *ring_no = AcToRing_GE1Map[lan_traffic][ac];
 	}
     }else { //GMAC2
-#if defined (CONFIG_RALINK_RT2883) 
+#if defined (CONFIG_RALINK_RT2883) && defined (CONFIG_RAETH_GMAC2)
 	*ring_no = AcToRing_GE2Map[ac];
 #endif
     }
 
 
     /* Set Port No - PN field in Tx Descriptor*/
-#if defined (CONFIG_RALINK_RT2883) 
+#if defined (CONFIG_RALINK_RT2883) && defined(CONFIG_RAETH_GMAC2)
     *port_no = gmac_no;
-#elif defined (CONFIG_RALINK_RT3052) 
+#elif defined (CONFIG_RALINK_RT3052) || (defined (CONFIG_RALINK_RT2883) && !defined(CONFIG_RAETH_GMAC2))
     if(bridge_traffic) {
 	*port_no = 1;
     }else {
