diff -urN SDK_3_1_0_0/RT288x_SDK/source/linux-2.6.21.x/drivers/net/raeth/raether.h SDK_3_2_0_0/RT288x_SDK/source/linux-2.6.21.x//drivers/net/raeth/raether.h
--- SDK_3_1_0_0/RT288x_SDK/source/linux-2.6.21.x/drivers/net/raeth/raether.h	2008-08-06 05:27:47.000000000 +0200
+++ SDK_3_2_0_0/RT288x_SDK/source/linux-2.6.21.x//drivers/net/raeth/raether.h	2008-10-24 08:20:14.000000000 +0200
@@ -15,15 +15,32 @@
 #else
 #define NUM_TX_RINGS 	4
 #endif
-
+#ifdef MEMORY_OPTIMIZATION
+#ifdef CONFIG_RAETH_ROUTER
+#define NUM_RX_DESC     128
+#define NUM_TX_DESC    	128
+#elif CONFIG_RT_3052_ESW
+#define NUM_RX_DESC     64
+#define NUM_TX_DESC     64
+#else
+#define NUM_RX_DESC     128
+#define NUM_TX_DESC     128
+#endif
+//#define NUM_RX_MAX_PROCESS 32
+#define NUM_RX_MAX_PROCESS 64
+#else
 #ifdef CONFIG_RAETH_ROUTER
 #define NUM_RX_DESC     256
 #define NUM_TX_DESC    	256
+#elif CONFIG_RT_3052_ESW
+#define NUM_RX_DESC     256
+#define NUM_TX_DESC     256
 #else
 #define NUM_RX_DESC     256
 #define NUM_TX_DESC     256
 #endif
 #define NUM_RX_MAX_PROCESS 64
+#endif
 
 #define DEV_NAME        "eth2"
 #define RA2882_ENET0	RA2882ETH_BASE
@@ -62,4 +79,7 @@
 
 void ei_xmit_housekeeping(unsigned long data);
 
+u32 mii_mgr_read(u32 phy_addr, u32 phy_register, u32 *read_data);
+u32 mii_mgr_write(u32 phy_addr, u32 phy_register, u32 write_data);
+
 #endif
