diff -Naurp -x brcm-boards -x pci linux-2.6.16.7-generic-patched/arch/mips/kernel/cpu-probe.c linux-2.6.16.7-patched/arch/mips/kernel/cpu-probe.c
--- linux-2.6.16.7-generic-patched/arch/mips/kernel/cpu-probe.c	2006-04-17 23:53:25.000000000 +0200
+++ linux-2.6.16.7-patched/arch/mips/kernel/cpu-probe.c	2006-07-05 15:21:58.000000000 +0200
@@ -558,6 +558,25 @@ static inline void decode_configs(struct
 		return;
 }
 
+static inline void cpu_probe_broadcom(struct cpuinfo_mips *c)
+{
+	decode_configs(c);
+	switch (c->processor_id & 0xff00) {
+	case PRID_IMP_BCM6338:          
+		c->cputype = CPU_BCM6338;
+		break;
+	case PRID_IMP_BCM6345:
+		c->cputype = CPU_BCM6345;
+		break;
+	case PRID_IMP_BCM6348:
+		c->cputype = CPU_BCM6348;
+		break;
+	default:
+		c->cputype = CPU_UNKNOWN;
+		break;
+	}
+}
+
 static inline void cpu_probe_mips(struct cpuinfo_mips *c)
 {
 	decode_configs(c);
@@ -691,6 +710,9 @@ __init void cpu_probe(void)
 	case PRID_COMP_LEGACY:
 		cpu_probe_legacy(c);
 		break;
+	case PRID_COMP_BROADCOM:
+		cpu_probe_broadcom(c);
+		break;
 	case PRID_COMP_MIPS:
 		cpu_probe_mips(c);
 		break;
