--- a/edit/syntax.c
+++ b/edit/syntax.c
@@ -1193,10 +1193,7 @@ edit_load_syntax (WEdit *edit, char ***p
     r = edit_read_syntax_file (edit, pnames, f, edit ? edit->filename : 0,
 			       get_first_editor_line (edit), type);
     if (r == -1) {
-	edit_free_syntax_rules (edit);
-	message (D_ERROR, _(" Load syntax file "),
-		 _(" Cannot open file %s \n %s "), f,
-		 unix_error_string (errno));
+	/* missing syntax file */
     } else if (r) {
 	edit_free_syntax_rules (edit);
 	message (D_ERROR, _(" Load syntax file "),
