diff -urN linux-2.6.21.1/include/linux/mtd/cfi.h linux-2.6.21.x/include/linux/mtd/cfi.h
--- linux-2.6.21.1/include/linux/mtd/cfi.h	2007-04-27 23:49:26.000000000 +0200
+++ linux-2.6.21.x/include/linux/mtd/cfi.h	2008-08-06 05:27:59.000000000 +0200
@@ -1,7 +1,7 @@
 
 /* Common Flash Interface structures
  * See http://support.intel.com/design/flash/technote/index.htm
- * $Id: cfi.h,v 1.57 2005/11/15 23:28:17 tpoynor Exp $
+ * $Id: cfi.h,v 1.2 2007-06-05 08:05:17 steven Exp $
  */
 
 #ifndef __MTD_CFI_H__
@@ -201,6 +201,27 @@
 
 /* Vendor-Specific PRI for Atmel chips (command set 0x0002) */
 
+/* Vendor-Specific PRI for SST Extended Command Set (0x0701) */
+
+struct cfi_pri_sststd {
+	uint8_t  pri[3];
+	uint8_t  MajorVersion;
+	uint8_t  MinorVersion;
+	uint8_t  SiliconRevision; /* bits 1-0: Address Sensitive Unlock */
+	uint8_t  EraseSuspend;
+	uint8_t  BlkProt;
+	uint8_t  TmpBlkUnprotect;
+	uint8_t  BlkProtUnprot;
+	uint8_t  SimultaneousOps;
+	uint8_t  BurstMode;
+	uint8_t  PageMode;
+	uint8_t  VppMin;
+	uint8_t  VppMax;
+	uint8_t  TopBottom;
+} __attribute__((packed));
+
+/* Vendor-Specific PRI for SST chips (command set 0x0701) */
+
 struct cfi_pri_atmel {
 	uint8_t pri[3];
 	uint8_t MajorVersion;
@@ -478,6 +499,7 @@
 #define CFI_MFR_AMD 0x0001
 #define CFI_MFR_ATMEL 0x001F
 #define CFI_MFR_ST  0x0020 	/* STMicroelectronics */
+#define CFI_MFR_SST 0x00BF
 
 void cfi_fixup(struct mtd_info *mtd, struct cfi_fixup* fixups);
 
