diff -Nur olsrd-0.4.10.orig/lib/httpinfo/Makefile olsrd-0.4.10/lib/httpinfo/Makefile
--- olsrd-0.4.10.orig/lib/httpinfo/Makefile	2005-05-25 15:50:22.000000000 +0200
+++ olsrd-0.4.10/lib/httpinfo/Makefile	2006-01-28 22:04:25.000000000 +0100
@@ -48,7 +48,11 @@
 	CFLAGS += -DADMIN_INTERFACE
 endif
 
+ifdef SVEN_OLA
+	CFLAGS += -DSVEN_OLA
+else
 OBJS += $(TOPDIR)/src/cfgparser/cfgfile_gen.o
+endif
 
 default_target: $(PLUGIN_FULLNAME)
 ifdef ADMIN_INTERFACE
diff -Nur olsrd-0.4.10.orig/lib/httpinfo/src/html.h olsrd-0.4.10/lib/httpinfo/src/html.h
--- olsrd-0.4.10.orig/lib/httpinfo/src/html.h	2005-03-14 22:28:15.000000000 +0100
+++ olsrd-0.4.10/lib/httpinfo/src/html.h	2006-01-28 22:04:25.000000000 +0100
@@ -90,11 +90,14 @@
   "<HEAD>\n",
   "<META http-equiv=\"Content-type\" content=\"text/html; charset=ISO-8859-1\">\n",
   "<TITLE>olsr.org httpinfo plugin</TITLE>\n",
+#ifndef SVEN_OLA
   "<link rel=\"icon\" href=\"images/favicon.ico\" type=\"image/x-icon\">\n",
   "<link rel=\"shortcut icon\" href=\"images/favicon.ico\" type=\"image/x-icon\">\n",
+#endif
   "<link rel=\"stylesheet\" type=\"text/css\" href=\"httpinfo.css\">\n",
   "</HEAD>\n",
   "<body bgcolor=\"#ffffff\" text=\"#000000\">\n",
+#ifndef SVEN_OLA
   "<table align=\"center\" border=\"0\" cellpadding=\"0\" cellspacing=\"0\" width=\"800\">"
   "<tbody><tr bgcolor=\"#ffffff\">",
   "<td align=\"left\" height=\"69\" valign=\"middle\" width=\"80%\">",
@@ -105,6 +108,7 @@
   "<p>",
   "</table>",
   "<!-- END HEAD -->\n\n",
+#endif
   NULL
 };
 
diff -Nur olsrd-0.4.10.orig/lib/httpinfo/src/olsrd_httpinfo.c olsrd-0.4.10/lib/httpinfo/src/olsrd_httpinfo.c
--- olsrd-0.4.10.orig/lib/httpinfo/src/olsrd_httpinfo.c	2005-12-22 16:06:52.000000000 +0100
+++ olsrd-0.4.10/lib/httpinfo/src/olsrd_httpinfo.c	2006-01-28 22:05:06.000000000 +0100
@@ -61,7 +61,9 @@
 #include "olsrd_httpinfo.h"
 #include "admin_interface.h"
 #include "html.h"
+#ifndef SVEN_OLA
 #include "gfx.h"
+#endif
 
 #ifdef OS
 #undef OS
@@ -206,10 +208,13 @@
     {"Admin", "admin", build_admin_body, OLSR_TRUE},
 #endif
     {"About", "about", build_about_body, OLSR_TRUE},
+#ifndef SVEN_OLA
     {"FOO", "cfgfile", build_cfgfile_body, OLSR_FALSE},
+#endif
     {NULL, NULL, NULL}
   };
 
+#ifndef SVEN_OLA
 struct static_bin_file_entry static_bin_files[] =
   {
     {"favicon.ico", favicon_ico, 1406/*favicon_ico_len*/},
@@ -217,6 +222,7 @@
     {"grayline.gif", grayline_gif, 43/*grayline_gif_len*/},
     {NULL, NULL, 0}
   };
+#endif
 
 struct static_txt_file_entry static_txt_files[] =
   {
@@ -418,6 +424,7 @@
       int i = 0;
       int y = 0;
 
+#ifndef SVEN_OLA
       while(static_bin_files[i].filename)
 	{
 	  if(FILENREQ_MATCH(filename, static_bin_files[i].filename))
@@ -435,6 +442,7 @@
 	}
 
       i = 0;
+#endif
 
       while(static_txt_files[i].filename)
 	{
@@ -1180,6 +1188,7 @@
   return size;
 }
 
+#ifndef SVEN_OLA
 int
 build_cfgfile_body(char *buf, olsr_u32_t bufsize)
 {
@@ -1207,7 +1216,7 @@
  
   return size;
 }
-
+#endif
 
 olsr_bool
 check_allowed_ip(union olsr_ip_addr *addr)
diff -Nur olsrd-0.4.10.orig/Makefile.inc olsrd-0.4.10/Makefile.inc
--- olsrd-0.4.10.orig/Makefile.inc	2006-01-01 16:58:20.000000000 +0100
+++ olsrd-0.4.10/Makefile.inc	2006-01-28 22:04:25.000000000 +0100
@@ -69,6 +69,10 @@
 # we always need the includes and defines
 CFLAGS += $(INCLUDES) $(DEFINES)
 
+ifdef SVEN_OLA
+CFLAGS += -DSVEN_OLA -DNODEBUG
+endif
+
 TAGFILE ?=	src/TAGS
 
 help:
diff -Nur olsrd-0.4.10.orig/src/cfgparser/Makefile olsrd-0.4.10/src/cfgparser/Makefile
--- olsrd-0.4.10.orig/src/cfgparser/Makefile	2005-12-29 20:48:43.000000000 +0100
+++ olsrd-0.4.10/src/cfgparser/Makefile	2006-01-28 22:04:25.000000000 +0100
@@ -41,7 +41,11 @@
 TOPDIR=../..
 include $(TOPDIR)/Makefile.inc
 
+ifdef SVEN_OLA
+SRCS =		olsrd_conf.c  oparse.c  oscan.c
+else
 SRCS =		olsrd_conf.c  oparse.c  oscan.c cfgfile_gen.c
+endif
 HDRS =		olsrd_conf.h  oparse.h
 
 ifeq ($(OS), win32)
diff -Nur olsrd-0.4.10.orig/src/defs.h olsrd-0.4.10/src/defs.h
--- olsrd-0.4.10.orig/src/defs.h	2006-01-01 16:59:02.000000000 +0100
+++ olsrd-0.4.10/src/defs.h	2006-01-28 22:04:25.000000000 +0100
@@ -176,6 +176,7 @@
 /* TC empty message sending */
 extern clock_t send_empty_tc;
 
+#ifndef SVEN_OLA
 /*
  *IPC functions
  *These are moved to a plugin soon
@@ -201,3 +202,4 @@
 ipc_route_send_rtentry(union olsr_ip_addr*, union olsr_ip_addr *, int, int, char *);
 
 #endif
+#endif
diff -Nur olsrd-0.4.10.orig/src/duplicate_set.c olsrd-0.4.10/src/duplicate_set.c
--- olsrd-0.4.10.orig/src/duplicate_set.c	2005-02-27 19:39:43.000000000 +0100
+++ olsrd-0.4.10/src/duplicate_set.c	2006-01-28 22:04:25.000000000 +0100
@@ -349,6 +349,7 @@
 
 
 
+#ifndef SVEN_OLA
 
 void
 olsr_print_duplicate_table()
@@ -373,3 +374,4 @@
 printf("\n");
 
 }
+#endif
diff -Nur olsrd-0.4.10.orig/src/hna_set.c olsrd-0.4.10/src/hna_set.c
--- olsrd-0.4.10.orig/src/hna_set.c	2005-05-29 14:47:45.000000000 +0200
+++ olsrd-0.4.10/src/hna_set.c	2006-01-28 22:04:25.000000000 +0100
@@ -318,6 +318,7 @@
  *
  *@return nada
  */
+#ifndef SVEN_OLA
 void
 olsr_print_hna_set()
 {
@@ -364,5 +365,6 @@
     }
 
 }
+#endif
 
 
diff -Nur olsrd-0.4.10.orig/src/ipc_frontend.c olsrd-0.4.10/src/ipc_frontend.c
--- olsrd-0.4.10.orig/src/ipc_frontend.c	2005-11-10 20:33:57.000000000 +0100
+++ olsrd-0.4.10/src/ipc_frontend.c	2006-01-28 22:04:25.000000000 +0100
@@ -46,6 +46,7 @@
  *
  */
 
+#ifndef SVEN_OLA
 #include "ipc_frontend.h"
 #include "link_set.h"
 #include "olsr.h"
@@ -541,3 +542,4 @@
   
   return 1;
 }
+#endif
diff -Nur olsrd-0.4.10.orig/src/link_set.c olsrd-0.4.10/src/link_set.c
--- olsrd-0.4.10.orig/src/link_set.c	2005-11-17 05:25:44.000000000 +0100
+++ olsrd-0.4.10/src/link_set.c	2006-01-28 22:04:25.000000000 +0100
@@ -849,6 +849,7 @@
   return;
 }
 
+#ifndef SVEN_OLA
 void olsr_print_link_set(void)
 {
   struct link_entry *walker;
@@ -889,6 +890,7 @@
 		walker->neigh_link_quality, etx)
   }
 }
+#endif
 
 static void update_packet_loss_worker(struct link_entry *entry, int lost)
 {
diff -Nur olsrd-0.4.10.orig/src/linux/kernel_routes.c olsrd-0.4.10/src/linux/kernel_routes.c
--- olsrd-0.4.10.orig/src/linux/kernel_routes.c	2005-02-27 19:39:43.000000000 +0100
+++ olsrd-0.4.10/src/linux/kernel_routes.c	2006-01-28 22:04:25.000000000 +0100
@@ -135,6 +135,7 @@
    *Send IPC route update message
    */
   
+#ifndef SVEN_OLA
   if(olsr_cnf->open_ipc)
       {
 	ipc_route_send_rtentry(&destination->rt_dst, 
@@ -143,6 +144,7 @@
 			       1,
 			       destination->rt_if->int_name); /* Send interface name */
       }
+#endif
   
   
   if (ifnet && kernel_route.rt_dev)
@@ -210,6 +212,7 @@
 
   if((tmp = ioctl(ioctl_s, SIOCADDRT, &kernel_route)) >= 0)
     {
+#ifndef SVEN_OLA
       if(olsr_cnf->open_ipc)
 	{
 	  if(memcmp(&destination->rt_router, &null_addr6, ipsize) != 0)
@@ -220,6 +223,7 @@
 				   destination->rt_if->int_name); /* Send interface name */
 
 	}
+#endif
     }
     return(tmp);
 }
@@ -281,12 +285,14 @@
      *Send IPC route update message
      */
 
+#ifndef SVEN_OLA
   if(olsr_cnf->open_ipc)
     ipc_route_send_rtentry(&destination->rt_dst, 
 			   NULL, 
 			   destination->rt_metric, 
 			   0,
 			   NULL); /* Send interface name */
+#endif
 
   return tmp;
 }
@@ -338,12 +344,14 @@
      *Send IPC route update message
      */
 
+#ifndef SVEN_OLA
   if(olsr_cnf->open_ipc)
     ipc_route_send_rtentry(&destination->rt_dst, 
 			   NULL, 
 			   destination->rt_metric, 
 			   0,
 			   NULL); /* Send interface name */
+#endif
 
   return tmp;
 }
diff -Nur olsrd-0.4.10.orig/src/lq_avl.c olsrd-0.4.10/src/lq_avl.c
--- olsrd-0.4.10.orig/src/lq_avl.c	2005-01-22 15:30:57.000000000 +0100
+++ olsrd-0.4.10/src/lq_avl.c	2006-01-28 22:04:25.000000000 +0100
@@ -57,6 +57,12 @@
 {
   int diff;
 
+#ifdef SVEN_OLA
+  if (0 == comp) {
+    diff = svenola_avl_comp_ipv4(key, node->key);
+  }
+  else
+#endif
   diff = (*comp)(key, node->key);
 
   if (diff < 0)
@@ -87,6 +93,13 @@
 
   node = avl_find_rec(tree->root, key, tree->comp);
 
+#ifdef SVEN_OLA
+  if (0 == tree->comp) {
+    if (0 != svenola_avl_comp_ipv4(node->key, key))
+      return NULL;
+  }
+  else
+#endif
   if ((*tree->comp)(node->key, key) != 0)
     return NULL;
 
@@ -228,6 +241,12 @@
 
   node = avl_find_rec(tree->root, new->key, tree->comp);
 
+#ifdef SVEN_OLA
+  if (0 == tree->comp) {
+    diff = svenola_avl_comp_ipv4(new->key, node->key);
+  }
+  else
+#endif
   diff = (*tree->comp)(new->key, node->key);
 
   if (diff == 0)
diff -Nur olsrd-0.4.10.orig/src/lq_avl.h olsrd-0.4.10/src/lq_avl.h
--- olsrd-0.4.10.orig/src/lq_avl.h	2005-02-20 19:52:18.000000000 +0100
+++ olsrd-0.4.10/src/lq_avl.h	2006-01-28 22:04:25.000000000 +0100
@@ -62,4 +62,9 @@
 struct avl_node *avl_find(struct avl_tree *, void *);
 int avl_insert(struct avl_tree *, struct avl_node *);
 
+#ifdef SVEN_OLA
+#define svenola_avl_comp_ipv4(ip1, ip2) \
+  (*(unsigned int *)ip1 == *(unsigned int *)ip2 ? 0 : \
+  *(unsigned int *)ip1 < *(unsigned int *)ip2 ? -1 : +1)
+#endif
 #endif
diff -Nur olsrd-0.4.10.orig/src/lq_list.c olsrd-0.4.10/src/lq_list.c
--- olsrd-0.4.10.orig/src/lq_list.c	2004-12-04 18:06:57.000000000 +0100
+++ olsrd-0.4.10/src/lq_list.c	2006-01-28 22:04:25.000000000 +0100
@@ -48,6 +48,7 @@
   list->tail = NULL;
 }
 
+#ifndef SVEN_OLA
 struct list_node *list_get_head(struct list *list)
 {
   return list->head;
@@ -67,6 +68,7 @@
 {
   return node->prev;
 }
+#endif
 
 void list_add_head(struct list *list, struct list_node *node)
 {
diff -Nur olsrd-0.4.10.orig/src/lq_list.h olsrd-0.4.10/src/lq_list.h
--- olsrd-0.4.10.orig/src/lq_list.h	2005-02-20 19:52:18.000000000 +0100
+++ olsrd-0.4.10/src/lq_list.h	2006-01-28 22:04:25.000000000 +0100
@@ -58,11 +58,18 @@
 
 void list_init(struct list *list);
 
+#ifdef SVEN_OLA
+#define list_get_head(node) (node)->head
+#define list_get_tail(node) (node)->tail
+#define list_get_next(node) (node)->next
+#define list_get_prev(node) (node)->prev
+#else
 struct list_node *list_get_head(struct list *list);
 struct list_node *list_get_tail(struct list *list);
 
 struct list_node *list_get_next(struct list_node *node);
 struct list_node *list_get_prev(struct list_node *node);
+#endif
 
 void list_add_head(struct list *list, struct list_node *node);
 void list_add_tail(struct list *list, struct list_node *node);
diff -Nur olsrd-0.4.10.orig/src/lq_route.c olsrd-0.4.10/src/lq_route.c
--- olsrd-0.4.10.orig/src/lq_route.c	2005-11-29 19:37:58.000000000 +0100
+++ olsrd-0.4.10/src/lq_route.c	2006-01-28 22:04:25.000000000 +0100
@@ -205,6 +205,16 @@
 
   // add the vertex to the list, if it's not us
 
+#ifdef SVEN_OLA
+  if (0 == comp) {
+    if (svenola_avl_comp_ipv4(&main_addr, node->key) != 0)
+    {
+      vert->node.data = vert;
+      list_add_tail(vertex_list, &vert->node);
+    }
+  }
+  else
+#endif
   if ((*comp)(&main_addr, node->key) != 0)
   {
     vert->node.data = vert;
@@ -371,7 +381,11 @@
   struct interface *inter;
 
   if (ipsize == 4)
+#ifdef SVEN_OLA
+    avl_comp = 0;
+#else
     avl_comp = avl_comp_ipv4;
+#endif
 
   else
     avl_comp = avl_comp_ipv6;
diff -Nur olsrd-0.4.10.orig/src/main.c olsrd-0.4.10/src/main.c
--- olsrd-0.4.10.orig/src/main.c	2005-09-29 07:53:34.000000000 +0200
+++ olsrd-0.4.10/src/main.c	2006-01-28 22:04:25.000000000 +0100
@@ -352,8 +352,10 @@
 
   /* Initialize the IPC socket */
 
+#ifndef SVEN_OLA
   if(olsr_cnf->open_ipc)
       ipc_init();
+#endif
 
   /* Initialisation of different tables to be used.*/
   olsr_init_tables();
@@ -465,9 +467,11 @@
 
   OLSR_PRINTF(1, "Closing sockets...\n")
 
+#ifndef SVEN_OLA
   /* front-end IPC socket */
   if(olsr_cnf->open_ipc)
     shutdown_ipc();
+#endif
 
   /* OLSR sockets */
   for (ifn = ifnet; ifn; ifn = ifn->int_next) 
@@ -528,7 +532,11 @@
   fprintf(stderr, "An error occured somwhere between your keyboard and your chair!\n"); 
   fprintf(stderr, "usage: olsrd [-f <configfile>] [ -i interface1 interface2 ... ]\n");
   fprintf(stderr, "  [-d <debug_level>] [-ipv6] [-multi <IPv6 multicast address>]\n"); 
+#ifdef SVEN_OLA
+  fprintf(stderr, "  [-bcast <broadcastaddr>] [-delgw] (Note: no -ipc,-dispin,-dispout)\n");
+#else
   fprintf(stderr, "  [-bcast <broadcastaddr>] [-ipc] [-dispin] [-dispout] [-delgw]\n");
+#endif
   fprintf(stderr, "  [-hint <hello interval (secs)>] [-tcint <tc interval (secs)>]\n");
   fprintf(stderr, "  [-midint <mid interval (secs)>] [-hnaint <hna interval (secs)>]\n");
   fprintf(stderr, "  [-T <Polling Rate (secs)>] [-nofork] [-hemu <ip_address>] \n"); 
@@ -735,6 +743,7 @@
 	}
 
 
+#ifndef SVEN_OLA
       /*
        * Should we display the contents of packages beeing sent?
        */
@@ -763,6 +772,7 @@
 	  cnf->open_ipc = OLSR_TRUE;
 	  continue;
 	}
+#endif
 
       /*
        * IPv6 multicast addr
diff -Nur olsrd-0.4.10.orig/src/neighbor_table.c olsrd-0.4.10/src/neighbor_table.c
--- olsrd-0.4.10.orig/src/neighbor_table.c	2005-11-29 19:37:58.000000000 +0100
+++ olsrd-0.4.10/src/neighbor_table.c	2006-01-28 22:04:25.000000000 +0100
@@ -468,6 +468,7 @@
  *
  *@return nada
  */
+#ifndef SVEN_OLA
 void
 olsr_print_neighbor_table()
 {
@@ -517,6 +518,7 @@
         }
     }
 }
+#endif
 
 
 
diff -Nur olsrd-0.4.10.orig/src/net_olsr.c olsrd-0.4.10/src/net_olsr.c
--- olsrd-0.4.10.orig/src/net_olsr.c	2005-12-29 19:37:16.000000000 +0100
+++ olsrd-0.4.10/src/net_olsr.c	2006-01-28 22:04:25.000000000 +0100
@@ -43,7 +43,9 @@
 #include "log.h"
 #include "olsr.h"
 #include "net_os.h"
+#ifndef SVEN_OLA
 #include "print_packet.h"
+#endif
 #include "link_set.h"
 #include <stdlib.h>
 
@@ -448,9 +450,11 @@
    *if the -dispout option was given
    *we print the contetnt of the packets
    */
+#ifndef SVEN_OLA
   if(disp_pack_out)
     print_olsr_serialized_packet(stdout, (union olsr_packet *)netbufs[ifp->if_nr]->buff, 
 				 netbufs[ifp->if_nr]->pending, &ifp->ip_addr); 
+#endif
   
   if(olsr_cnf->ip_version == AF_INET)
     {
diff -Nur olsrd-0.4.10.orig/src/olsr.c olsrd-0.4.10/src/olsr.c
--- olsrd-0.4.10.orig/src/olsr.c	2005-11-17 05:25:44.000000000 +0100
+++ olsrd-0.4.10/src/olsr.c	2006-01-28 22:04:25.000000000 +0100
@@ -209,6 +209,7 @@
         }
     }
   
+#ifndef SVEN_OLA
   if (olsr_cnf->debug_level > 0)
     {      
       if (olsr_cnf->debug_level > 2) 
@@ -226,6 +227,7 @@
       olsr_print_neighbor_table();
       olsr_print_tc_table();
     }
+#endif
 
   for(tmp_pc_list = pcf_list; 
       tmp_pc_list != NULL;
diff -Nur olsrd-0.4.10.orig/src/parser.c olsrd-0.4.10/src/parser.c
--- olsrd-0.4.10.orig/src/parser.c	2005-11-19 09:49:44.000000000 +0100
+++ olsrd-0.4.10/src/parser.c	2006-01-28 22:04:25.000000000 +0100
@@ -50,7 +50,9 @@
 #include "rebuild_packet.h"
 #include "net_os.h"
 #include "log.h"
+#ifndef SVEN_OLA
 #include "print_packet.h"
+#endif
 
 #ifdef WIN32
 #undef EWOULDBLOCK
@@ -174,9 +176,11 @@
 
   //printf("Message from %s\n\n", olsr_ip_to_string(from_addr)); 
       
+#ifndef SVEN_OLA
   /* Display packet */
   if(disp_pack_in)
     print_olsr_serialized_packet(stdout, (union olsr_packet *)olsr, size, from_addr);
+#endif
 
   if(olsr_cnf->ip_version == AF_INET)
     msgsize = ntohs(m->v4.olsr_msgsize);
diff -Nur olsrd-0.4.10.orig/src/print_packet.c olsrd-0.4.10/src/print_packet.c
--- olsrd-0.4.10.orig/src/print_packet.c	2005-11-19 09:49:44.000000000 +0100
+++ olsrd-0.4.10/src/print_packet.c	2006-01-28 22:04:25.000000000 +0100
@@ -39,6 +39,7 @@
  * $Id: olsrd-optimize.patch,v 1.4 2006/01/28 21:21:08 sven-ola Exp $
  */
 
+#ifndef SVEN_OLA
 #include "print_packet.h"
 #include "mantissa.h"
 #include "defs.h"
@@ -396,3 +397,4 @@
       remsize -= ipsize;
     }
 }
+#endif
diff -Nur olsrd-0.4.10.orig/src/tc_set.c olsrd-0.4.10/src/tc_set.c
--- olsrd-0.4.10.orig/src/tc_set.c	2005-11-17 05:25:44.000000000 +0100
+++ olsrd-0.4.10/src/tc_set.c	2006-01-28 22:04:25.000000000 +0100
@@ -430,6 +430,7 @@
 /**
  *Print the topology table to stdout
  */
+#ifndef SVEN_OLA
 int
 olsr_print_tc_table()
 {
@@ -489,3 +490,4 @@
 
   return 1;
 }
+#endif
diff -Nur olsrd-0.4.10.orig/src/two_hop_neighbor_table.c olsrd-0.4.10/src/two_hop_neighbor_table.c
--- olsrd-0.4.10.orig/src/two_hop_neighbor_table.c	2005-05-29 14:47:46.000000000 +0200
+++ olsrd-0.4.10/src/two_hop_neighbor_table.c	2006-01-28 22:04:25.000000000 +0100
@@ -236,6 +236,7 @@
 
 
 
+#ifndef SVEN_OLA
 /**
  *Print the two hop neighbor table to STDOUT.
  *
@@ -288,3 +289,4 @@
 	}
     }
 }
+#endif
diff -Nur olsrd-0.4.10.orig/src/win32/kernel_routes.c olsrd-0.4.10/src/win32/kernel_routes.c
--- olsrd-0.4.10.orig/src/win32/kernel_routes.c	2005-02-27 11:48:05.000000000 +0100
+++ olsrd-0.4.10/src/win32/kernel_routes.c	2006-01-28 22:04:25.000000000 +0100
@@ -94,9 +94,11 @@
     return -1;
   }
 
+#ifndef SVEN_OLA
   if(olsr_cnf->open_ipc)
     ipc_route_send_rtentry(&Dest->rt_dst, &Dest->rt_router, Dest->rt_metric,
                            1, Dest->rt_if->int_name);
+#endif
 
   return 0;
 }
@@ -144,8 +146,10 @@
     return -1;
   }
 
+#ifndef SVEN_OLA
   if(olsr_cnf->open_ipc)
     ipc_route_send_rtentry(&Dest->rt_dst, NULL, Dest->rt_metric, 0, NULL);
+#endif
 
   return 0;
 }
