diff -Nur olsrd-current.orig/src/defs.h olsrd-current/src/defs.h
--- olsrd-current.orig/src/defs.h	2006-01-04 16:51:19.000000000 +0100
+++ olsrd-current/src/defs.h	2006-01-07 09:19:51.000000000 +0100
@@ -71,7 +71,7 @@
 #define	HOPCNT_MAX		32	/* maximum hops number */
 #define	MAXMESSAGESIZE		1500	/* max broadcast size */
 #define UDP_IPV4_HDRSIZE        28
-#define UDP_IPV6_HDRSIZE        48
+#define UDP_IPV6_HDRSIZE        62
 #define MAX_IFS                 16
 
 /* Debug helper macro */
