Index: samba-3.0.24/source/Makefile
===================================================================
--- samba-3.0.24.orig/source/Makefile	2008-08-06 23:33:20.000000000 +0200
+++ samba-3.0.24/source/Makefile	2008-08-06 23:34:48.000000000 +0200
@@ -12,7 +12,7 @@
 
 # AR7
 # prefix=/usr/local/samba
-prefix=/var/samba
+prefix=
 exec_prefix=${prefix}
 
 # AR7
@@ -108,11 +108,11 @@
 # These can be overridden by command line switches (see smbd(8))
 # or in smb.conf (see smb.conf(5))
 LOGFILEBASE = ${VARDIR}
-CONFIGFILE = $(CONFIGDIR)/smb.conf
-LMHOSTSFILE = $(CONFIGDIR)/lmhosts
+CONFIGFILE = /etc/samba/smb.conf
+LMHOSTSFILE = /etc/lmhosts
 
 # This is where smbpasswd et al go
-PRIVATEDIR = ${prefix}/private
+PRIVATEDIR = /etc/samba/
 
 SMB_PASSWD_FILE = $(PRIVATEDIR)/smbpasswd
 PRIVATE_DIR = $(PRIVATEDIR)
@@ -121,10 +121,10 @@
 SWATDIR = ${prefix}/swat
 
 # the directory where lock files go
-LOCKDIR = ${VARDIR}/locks
+LOCKDIR = ${VARDIR}/lock
 
 # the directory where pid files go
-PIDDIR = ${VARDIR}/locks
+PIDDIR = ${VARDIR}/lock
 
 LIBSMBCLIENT=bin/libsmbclient.a bin/libsmbclient.so
 LIBSMBCLIENT_MAJOR=0
Index: samba-3.0.24/source/lib/util.c
===================================================================
--- samba-3.0.24.orig/source/lib/util.c	2008-08-06 23:41:58.000000000 +0200
+++ samba-3.0.24/source/lib/util.c	2008-08-06 23:42:10.000000000 +0200
@@ -300,11 +300,7 @@
 	char *p;
 	if ((p = getenv("TMPDIR")))
 		return p;
-#if 1 /* AR7 */
-	return "/var/tmp";
-#else
 	return "/tmp";
-#endif
 }
 
 /****************************************************************************
