Index: samba-3.0.24/source/include/config.h
===================================================================
--- samba-3.0.24.orig/source/include/config.h	2008-08-08 17:43:08.000000000 +0200
+++ samba-3.0.24/source/include/config.h	2008-08-08 17:44:02.000000000 +0200
@@ -821,7 +821,7 @@
 #define HAVE_MKTIME 1
 
 /* Whether mmap works */
-#define HAVE_MMAP 1
+/* #define HAVE_MMAP 1 */
 
 /* Define to 1 if you have the <mntent.h> header file. */
 #define HAVE_MNTENT_H 1
Index: samba-3.0.24/source/tdb/Makefile
===================================================================
--- samba-3.0.24.orig/source/tdb/Makefile	2008-08-08 17:42:47.000000000 +0200
+++ samba-3.0.24/source/tdb/Makefile	2008-08-08 17:44:15.000000000 +0200
@@ -2,7 +2,8 @@
 # Makefile for tdb directory
 #
 
-CFLAGS = -DSTANDALONE -DTDB_DEBUG -g -DHAVE_MMAP=1
+CFLAGS = -DSTANDALONE -DTDB_DEBUG -g 
+#-DHAVE_MMAP=1
 CC = gcc
 
 ADMINPROGS = tdbdump tdbbackup
