diff --git a/package/kernel/shortcut-fe/src/fast-classifier.c b/package/kernel/shortcut-fe/src/fast-classifier.c
index 92beef5a7b..e91f6f1043 100644
--- a/package/kernel/shortcut-fe/src/fast-classifier.c
+++ b/package/kernel/shortcut-fe/src/fast-classifier.c
@@ -35,6 +35,7 @@
 #include <linux/spinlock.h>
 #include <linux/if_bridge.h>
 #include <linux/hashtable.h>
+#include <net/pkt_sched.h>
 
 #include "sfe_backport.h"
 #include "sfe.h"
@@ -229,6 +230,16 @@ static int fast_classifier_recv(struct sk_buff *skb)
 		dev = master_dev;
 	}
 
+#ifdef CONFIG_NET_CLS_ACT
+	/*
+	 * If ingress Qdisc configured, and packet not processed by ingress Qdisc yet
+	 * We cannot accelerate this packet.
+	 */
+	if (dev->ingress_queue && !(skb->tc_verd & TC_NCLS)) {
+		goto rx_exit;
+	}
+#endif
+
 	/*
 	 * We're only interested in IPv4 and IPv6 packets.
 	 */
diff --git a/package/kernel/shortcut-fe/src/sfe_cm.c b/package/kernel/shortcut-fe/src/sfe_cm.c
index 2ef6f73e26..6d3e21085a 100644
--- a/package/kernel/shortcut-fe/src/sfe_cm.c
+++ b/package/kernel/shortcut-fe/src/sfe_cm.c
@@ -32,6 +32,7 @@
 #include <net/netfilter/nf_conntrack_timeout.h>
 #include <linux/netfilter/xt_dscp.h>
 #include <linux/if_bridge.h>
+#include <net/pkt_sched.h>
 
 #include "sfe.h"
 #include "sfe_cm.h"
@@ -135,6 +136,16 @@ static int sfe_cm_recv(struct sk_buff *skb)
 
 	dev = skb->dev;
 
+#ifdef CONFIG_NET_CLS_ACT
+	/*
+	 * If ingress Qdisc configured, and packet not processed by ingress Qdisc yet
+	 * We cannot accelerate this packet.
+	 */
+	if (dev->ingress_queue && !(skb->tc_verd & TC_NCLS)) {
+		return 0;
+	}
+#endif
+
 	/*
 	 * We're only interested in IPv4 and IPv6 packets.
 	 */
