diff -puNrb linux-2.6.27.39/arch/powerpc/include/asm/mtvic.h linux/arch/powerpc/include/asm/mtvic.h
--- linux-2.6.27.39/arch/powerpc/include/asm/mtvic.h	1970-01-01 03:00:00.000000000 +0300
+++ linux/arch/powerpc/include/asm/mtvic.h	2009-11-20 09:55:48.000000000 +0200
@@ -0,0 +1,8 @@
+#ifndef _ASM_POWERPC_MTVIC_H
+#define _ASM_POWERPC_MTVIC_H
+
+void __init mtvic_init(int def);
+unsigned mtvic_get_irq(void);
+unsigned rb_get_irq(void);
+
+#endif
