diff -puNrb linux-2.6.22.14/drivers/char/random.c linux/drivers/char/random.c
--- linux-2.6.22.14/drivers/char/random.c	2008-01-22 14:02:19.000000000 +0200
+++ linux/drivers/char/random.c	2008-01-15 11:03:15.000000000 +0200
@@ -1563,6 +1563,7 @@ __u32 secure_tcp_sequence_number(__be32 
 #endif
 	return seq;
 }
+EXPORT_SYMBOL_GPL(secure_tcp_sequence_number);
 
 /* Generate secure starting point for ephemeral IPV4 transport port search */
 u32 secure_ipv4_port_ephemeral(__be32 saddr, __be32 daddr, __be16 dport)
