diff -Naurp -x brcm-boards -x pci linux-2.6.16.7-generic-patched/arch/mips/mm/tlbex.c linux-2.6.16.7-patched/arch/mips/mm/tlbex.c
--- linux-2.6.16.7-generic-patched/arch/mips/mm/tlbex.c	2006-04-17 23:53:25.000000000 +0200
+++ linux-2.6.16.7-patched/arch/mips/mm/tlbex.c	2006-07-05 15:21:58.000000000 +0200
@@ -858,6 +858,9 @@ static __init void build_tlb_write_entry
 	case CPU_4KSC:
 	case CPU_20KC:
 	case CPU_25KF:
+	case CPU_BCM6338:
+	case CPU_BCM6345:
+	case CPU_BCM6348:
 		tlbw(p);
 		break;
 
