diff -puNrb linux-2.6.22.14/mm/page_alloc.c linux/mm/page_alloc.c
--- linux-2.6.22.14/mm/page_alloc.c	2008-01-22 13:58:54.000000000 +0200
+++ linux/mm/page_alloc.c	2008-01-15 10:59:27.000000000 +0200
@@ -1239,6 +1239,14 @@ restart:
 		goto got_pg;
 
 	/*
+	 * Code in arch/mips/kernel/module.c wants physically
+	 * contiguous memory only if there is plenty of free of them.
+	 */
+	if ((gfp_mask & (__GFP_THISNODE | __GFP_NORETRY | __GFP_NOWARN))
+	    == (__GFP_THISNODE | __GFP_NORETRY | __GFP_NOWARN))
+		goto nopage;
+
+	/*
 	 * GFP_THISNODE (meaning __GFP_THISNODE, __GFP_NORETRY and
 	 * __GFP_NOWARN set) should not cause reclaim since the subsystem
 	 * (f.e. slab) using GFP_THISNODE may choose to trigger reclaim
@@ -1252,6 +1260,9 @@ restart:
 	for (z = zonelist->zones; *z; z++)
 		wakeup_kswapd(*z, order);
 
+	if (gfp_mask & 0x80000000)
+	    goto nopage;
+
 	/*
 	 * OK, we're below the kswapd watermark and have kicked background
 	 * reclaim. Now things get more complex, so set up alloc_flags according
