diff -puNrb linux-2.6.22.14/mm/slub.c linux/mm/slub.c
--- linux-2.6.22.14/mm/slub.c	2008-01-22 13:58:54.000000000 +0200
+++ linux/mm/slub.c	2008-01-15 10:59:27.000000000 +0200
@@ -2199,7 +2199,7 @@ static struct kmem_cache *get_slab(size_
 		if (!x)
 			panic("Unable to allocate memory for dma cache\n");
 
-		if (index <= KMALLOC_SHIFT_HIGH)
+		if (index >= KMALLOC_SHIFT_LOW)
 			realsize = 1 << index;
 		else {
 			if (index == 1)
