--- test/tester.tcl
+++ test/tester.tcl
@@ -1441,6 +1441,7 @@
     set f [string range $f2 3 end]
     file copy -force $f2 $f
   }
+  file delete test.db-journal/__db.register
 }
 proc db_restore_and_reopen {{dbfile test.db}} {
   catch { db close }
@@ -1449,7 +1450,7 @@
 }
 proc db_delete_and_reopen {{file test.db}} {
   catch { db close }
-  foreach f [glob -nocomplain test.db*] { file delete -force $f }
+  foreach f [glob -nocomplain test.db*] { forcedelete $f }
   sqlite3 db $file
 }
 
--- test/permutations.test
+++ test/permutations.test
@@ -171,4 +171,9 @@
 }
 
+# We exlcude below fts3 tests:
+#  fts3defer.test, fts3defer2.test
+#  --  Known difference of zeroblob between SQLite and BDBSQL. #19764
+#  fts3fault.test
+#  --  Known difference: DBSQL testfixture does't support sqlite_io_error tests.
 test_suite "fts3" -prefix "" -description {
   All FTS3 tests except fts3rnd.test.
@@ -178,8 +183,8 @@
   fts3ak.test fts3al.test fts3am.test fts3an.test fts3ao.test
   fts3atoken.test fts3b.test fts3c.test fts3cov.test fts3d.test
-  fts3defer.test fts3defer2.test fts3e.test fts3expr.test fts3expr2.test 
+  fts3e.test fts3expr.test fts3expr2.test 
   fts3near.test fts3query.test fts3shared.test fts3snippet.test 
 
-  fts3fault.test fts3malloc.test fts3matchinfo.test
+  fts3malloc.test fts3matchinfo.test
 
   fts3aux1.test fts3comp1.test

