diff -Naurp -x mach-bcm963xx linux-2.6.16.7-generic-patched/include/asm-mips/mach-generic/param.h linux-2.6.16.7-patched/include/asm-mips/mach-generic/param.h
--- linux-2.6.16.7-generic-patched/include/asm-mips/mach-generic/param.h	2006-04-17 23:53:25.000000000 +0200
+++ linux-2.6.16.7-patched/include/asm-mips/mach-generic/param.h	2006-07-05 15:21:58.000000000 +0200
@@ -8,6 +8,6 @@
 #ifndef __ASM_MACH_GENERIC_PARAM_H
 #define __ASM_MACH_GENERIC_PARAM_H
 
-#define HZ		1000		/* Internal kernel timer frequency */
+#define HZ		200		/* Internal kernel timer frequency */
 
 #endif /* __ASM_MACH_GENERIC_PARAM_H */
