--- a/include/linux/init.h	2008-09-08 20:40:20.000000000 +0300
+++ b/include/linux/init.h	2008-11-07 10:45:15.000000000 +0200
@@ -128,6 +128,11 @@
 #define __REFDATA        .section       ".ref.data", "aw"
 #define __REFCONST       .section       ".ref.rodata", "aw"
 
+/* define kernel paramter variable for VS OpenRISC system ID */
+#if defined(CONFIG_ARCH_KS8695_VSOPENRISC) && !defined(__ASSEMBLY__)
+extern int vs_sysid;
+#endif 
+
 #ifndef __ASSEMBLY__
 /*
  * Used for initialization calls..
