--- a/include/linux/serial.h	2008-09-08 20:40:20.000000000 +0300
+++ b/include/linux/serial.h	2008-11-07 10:45:14.000000000 +0200
@@ -142,6 +142,7 @@
 #define ASYNC_CONS_FLOW		0x00800000 /* flow control for console  */
 
 #define ASYNC_BOOT_ONLYMCA	0x00400000 /* Probe only if MCA bus */
+#define ASYNC_VSOPENRISC_EPLD	0x00200000 /* usage of the EPLD's from the OpenRISC */
 #define ASYNC_INTERNAL_FLAGS	0xFFC00000 /* Internal flags */
 
 /*
