Index: core.c
===================================================================
--- core.c	(revision 5991)
+++ core.c	(working copy)
@@ -554,7 +554,7 @@
 		priv->band_24.n_bitrates = ARRAY_SIZE(mwl_rates_24);
 
 		mwl_set_ht_caps(priv, &priv->band_24);
-		mwl_set_vht_caps(priv, &priv->band_24, qam256);
+//		mwl_set_vht_caps(priv, &priv->band_24, qam256);
 
 		hw->wiphy->bands[NL80211_BAND_2GHZ] = &priv->band_24;
 	}
Index: hif/fwcmd.c
===================================================================
--- hif/fwcmd.c	(revision 5991)
+++ hif/fwcmd.c	(working copy)
@@ -1140,7 +1140,8 @@
 	u16 maxtxpow[SYSADPT_TX_GRP_PWR_LEVEL_TOTAL];
 	int i, tmp;
 	int rc = 0;
-
+	return rc;
+	
 	if ((priv->chip_type != MWL8997) && (priv->forbidden_setting))
 		return rc;
 
@@ -1188,6 +1189,13 @@
 		else
 			sub_ch = EXT_CH_BELOW_CTRL_CH;
 		break;
+	case NL80211_CHAN_WIDTH_160:
+		width = CH_160_MHZ_WIDTH;
+		if (conf->chandef.center_freq1 > channel->center_freq)
+			sub_ch = EXT_CH_ABOVE_CTRL_CH;
+		else
+			sub_ch = EXT_CH_BELOW_CTRL_CH;
+		break;
 	default:
 		return -EINVAL;
 	}
@@ -1249,6 +1257,8 @@
 	int index, found = 0;
 	int i, tmp;
 	int rc = 0;
+	return rc;
+	
 
 	if ((priv->chip_type != MWL8997) && (priv->forbidden_setting))
 		return rc;
@@ -1297,6 +1307,13 @@
 		else
 			sub_ch = EXT_CH_BELOW_CTRL_CH;
 		break;
+	case NL80211_CHAN_WIDTH_160:
+		width = CH_160_MHZ_WIDTH;
+		if (conf->chandef.center_freq1 > channel->center_freq)
+			sub_ch = EXT_CH_ABOVE_CTRL_CH;
+		else
+			sub_ch = EXT_CH_BELOW_CTRL_CH;
+		break;
 	default:
 		return -EINVAL;
 	}
@@ -3161,7 +3178,7 @@
 {
 	struct mwl_priv *priv = hw->priv;
 	struct hostcmd_cmd_set_cdd *pcmd;
-
+	priv->cdd = 1;
 	pcmd = (struct hostcmd_cmd_set_cdd *)&priv->pcmd_buf[0];
 
 	mutex_lock(&priv->fwcmd_mutex);
@@ -3437,6 +3454,7 @@
 	struct hostcmd_cmd_get_fw_region_code_sc4 *pcmd;
 	u16 cmd;
 
+	return -EIO;
 	pcmd = (struct hostcmd_cmd_get_fw_region_code_sc4 *)&priv->pcmd_buf[0];
 
 	mutex_lock(&priv->fwcmd_mutex);
@@ -3477,6 +3495,7 @@
 	int status;
 	u16 cmd;
 
+	return -EIO;
 	pcmd = (struct hostcmd_cmd_get_device_pwr_tbl_sc4 *)&priv->pcmd_buf[0];
 
 	mutex_lock(&priv->fwcmd_mutex);
