diff -urN freeradius-server-2.1.1/src/modules/rlm_ldap/configure freeradius-server-2.1.1.new/src/modules/rlm_ldap/configure
--- freeradius-server-2.1.1/src/modules/rlm_ldap/configure	2008-09-25 10:41:26.000000000 +0200
+++ freeradius-server-2.1.1.new/src/modules/rlm_ldap/configure	2008-10-13 06:29:23.000000000 +0200
@@ -3744,8 +3744,7 @@
 if test x"$fail" != x""; then
 	if test x"${enable_strict_dependencies}" = x"yes"; then
 		{ { echo "$as_me:$LINENO: error: set --without-rlm_ldap to disable it explicitly." >&5
-echo "$as_me: error: set --without-rlm_ldap to disable it explicitly." >&2;}
-   { (exit 1); exit 1; }; }
+echo "$as_me: error: set --without-rlm_ldap to disable it explicitly." >&2;} }
 	else
 		{ echo "$as_me:$LINENO: WARNING: silently not building rlm_ldap." >&5
 echo "$as_me: WARNING: silently not building rlm_ldap." >&2;}
