Index: ar2313.c
===================================================================
--- ar2313.c	(revision 11903)
+++ ar2313.c	(working copy)
@@ -931,8 +931,7 @@
 		printk("RX descr  %08x\n", rxdesc->descr);
 #endif
 
-		if ((status & (DMA_RX_ERROR | DMA_RX_ERR_LENGTH)) &&
-			(!(status & DMA_RX_LONG))) {
+		if ((status & DMA_RX_ERROR)){
 #if DEBUG_RX
 			printk("%s: rx ERROR %08x\n", __FUNCTION__, status);
 #endif
