diff -urN 1.4.0.94/linux/drivers/net/ag7100/athrs16_phy.h 1.4.0.144/linux/drivers/net/ag7100/athrs16_phy.h
--- 1.4.0.94/linux/drivers/net/ag7100/athrs16_phy.h	2009-06-11 20:46:37.000000000 +0200
+++ 1.4.0.144/linux/drivers/net/ag7100/athrs16_phy.h	2009-12-31 11:26:44.000000000 +0100
@@ -80,7 +80,7 @@
 #endif
 
 /*add feature define here*/
-//#define FULL_FEATURE
+#define FULL_FEATURE
 
 #ifdef CONFIG_ATHRS16_PHY 
 #undef HEADER_REG_CONF
