diff -urN SDK_3_1_0_0/RT288x_SDK/source/linux-2.6.21.x/drivers/flash/flash_ioctl.c SDK_3_2_0_0/RT288x_SDK/source/linux-2.6.21.x//drivers/flash/flash_ioctl.c
--- SDK_3_1_0_0/RT288x_SDK/source/linux-2.6.21.x/drivers/flash/flash_ioctl.c	2008-08-06 05:27:41.000000000 +0200
+++ SDK_3_2_0_0/RT288x_SDK/source/linux-2.6.21.x//drivers/flash/flash_ioctl.c	2008-10-24 08:20:14.000000000 +0200
@@ -26,11 +26,12 @@
                      unsigned int cmd, unsigned long arg)
 {
     struct flash_opt *opt=(struct flash_opt *)arg;
-    struct mii_mgr_opt *opt2=(struct mii_mgr_opt *)arg;
-    static char buf[FLASH_MAX_RW_SIZE];
+    unsigned char *buf;
     unsigned int tmp;
     unsigned int start_sect=0,end_sect=0;
 
+    buf=kmalloc(FLASH_MAX_RW_SIZE, GFP_KERNEL);
+
     switch(cmd) 
     {
     case FLASH_IOCTL_READ:
@@ -53,25 +54,11 @@
 		}
 	}
 	break;
-    case FLASH_IOCTL_MII_MGR_GET:
-	//printk("FLASH_IOCTL_MII_MGR_GET enter\n");
-	mii_mgr_read(opt2->phy_addr, opt2->reg_addr, &tmp);
-	copy_to_user(opt2->fromvalue, &tmp, 4);
-	//printk("Get: phy[%d].reg[%d] = %04x\n", opt2->phy_addr, opt2->reg_addr, tmp);
-	//printk("FLASH_IOCTL_MII_MGR_GET exit\n");
-	break;
-    case FLASH_IOCTL_MII_MGR_SET:
-	//printk("FLASH_IOCTL_MII_MGR_SET enter\n");
-	mii_mgr_write(opt2->phy_addr, opt2->reg_addr, opt2->tovalue );
-	mii_mgr_read(opt2->phy_addr, opt2->reg_addr, &tmp);
-	copy_to_user( opt2->fromvalue, &tmp, 4);
-	//printk("Get: phy[%d].reg[%d] = %04x\n", opt2->phy_addr, opt2->reg_addr, tmp);
-	//printk("FLASH_IOCTL_MII_MGR_SET exit\n");
-	break;
     default:
 	break;
     }
 
+    kfree(buf);
     return 0;
 }
 
