diff -urN SDK_3_1_0_0/RT288x_SDK/source/linux-2.6.21.x/drivers/flash/flash_tbl.h SDK_3_2_0_0/RT288x_SDK/source/linux-2.6.21.x//drivers/flash/flash_tbl.h
--- SDK_3_1_0_0/RT288x_SDK/source/linux-2.6.21.x/drivers/flash/flash_tbl.h	2008-08-06 05:27:41.000000000 +0200
+++ SDK_3_2_0_0/RT288x_SDK/source/linux-2.6.21.x//drivers/flash/flash_tbl.h	2008-10-24 08:20:14.000000000 +0200
@@ -18,16 +18,25 @@
 
 static struct flsh_dev flsh_tbl[] = 
 {
-#ifdef CONFIG_FLASH_MX29LV320
- { .man_id= 0xc2, /* Mxic */
-   .dev_id= 0x22A7, /* MX29LV320T */
+#ifdef CONFIG_FLASH_F49L320
+ { .man_id= 0x8C, /* ESMT */
+   .dev_id= 0x22F9, /* F49L320BA */
+   .max_sector = 70,
+   .sector= {
+             {.size=KBYTES(8),.count=8},
+	     {.size=KBYTES(64),.count=63}
+	    }
+ },
+ { .man_id= 0x8C, /* ESMT */
+   .dev_id= 0x22F6, /* F49L320UA */
    .max_sector = 70,
    .sector= {
 	     {.size=KBYTES(64),.count=63},
              {.size=KBYTES(8),.count=8}
 	    }
  },
-
+#endif
+#ifdef CONFIG_FLASH_MX29LV320
  { .man_id= 0xc2, /* Mxic */
    .dev_id= 0x22A8, /* MX29LV320B */
    .max_sector = 70,
@@ -116,7 +125,7 @@
 	    }
  },
 #endif
-#ifdef CONFIG_FLASH_EN29LV160A
+#ifdef CONFIG_FLASH_EN29LV160
  { .man_id= 0x7F, /* EON */
    .dev_id= 0x22C4, /* EN29LV160AT */
    .max_sector = 34,
