diff -urN linux-2.6.17.13/drivers/ide/ide.c linux-2.6.17/drivers/ide/ide.c
--- linux-2.6.17.13/drivers/ide/ide.c	2006-09-09 05:23:25.000000000 +0200
+++ linux-2.6.17/drivers/ide/ide.c	2006-11-08 01:10:27.983654500 +0100
@@ -1840,6 +1840,13 @@
 #ifdef CONFIG_H8300
 	h8300_ide_init();
 #endif
+#ifdef CONFIG_BLK_DEV_MAGICBOX_IDE
+	{
+		extern void ide_magicbox_init();
+		ide_magicbox_init();
+	}
+#endif
+
 }
 
 void ide_register_subdriver(ide_drive_t *drive, ide_driver_t *driver)
@@ -1847,6 +1854,7 @@
 #ifdef CONFIG_PROC_FS
 	ide_add_proc_entries(drive->proc, driver->proc, drive);
 #endif
+
 }
 
 EXPORT_SYMBOL(ide_register_subdriver);
