--- linux-6.1.57/include/net/netfilter/nf_conntrack_ecache.h	2023-10-11 03:00:46.000000000 +0700
+++ linux-6.1/include/net/netfilter/nf_conntrack_ecache.h	2023-10-01 15:59:14.476418884 +0700
@@ -60,14 +60,8 @@ struct nf_exp_event {
 	int report;
 };
 
-struct nf_ct_event_notifier {
-	int (*ct_event)(unsigned int events, const struct nf_ct_event *item);
-	int (*exp_event)(unsigned int events, const struct nf_exp_event *item);
-};
-
-void nf_conntrack_register_notifier(struct net *net,
-				   const struct nf_ct_event_notifier *nb);
-void nf_conntrack_unregister_notifier(struct net *net);
+extern int nf_conntrack_register_notifier(struct net *net, struct notifier_block *nb);
+extern int nf_conntrack_unregister_notifier(struct net *net, struct notifier_block *nb);
 
 void nf_ct_deliver_cached_events(struct nf_conn *ct);
 int nf_conntrack_eventmask_report(unsigned int eventmask, struct nf_conn *ct,
@@ -98,12 +92,8 @@ static inline void
 nf_conntrack_event_cache(enum ip_conntrack_events event, struct nf_conn *ct)
 {
 #ifdef CONFIG_NF_CONNTRACK_EVENTS
-	struct net *net = nf_ct_net(ct);
 	struct nf_conntrack_ecache *e;
 
-	if (!rcu_access_pointer(net->ct.nf_conntrack_event_cb))
-		return;
-
 	e = nf_ct_ecache_find(ct);
 	if (e == NULL)
 		return;
