diff -puNrb linux-2.6.22.14/include/asm-powerpc/processor.h linux/include/asm-powerpc/processor.h
--- linux-2.6.22.14/include/asm-powerpc/processor.h	2008-01-22 13:56:30.000000000 +0200
+++ linux/include/asm-powerpc/processor.h	2008-01-15 10:55:55.000000000 +0200
@@ -134,7 +134,7 @@ struct thread_struct {
 #ifdef CONFIG_PPC32
 	void		*pgdir;		/* root of page-table tree */
 #endif
-#if defined(CONFIG_4xx) || defined (CONFIG_BOOKE)
+#if defined(CONFIG_4xx) || defined (CONFIG_BOOKE) || defined(CONFIG_RB_PPC)
 	unsigned long	dbcr0;		/* debug control register values */
 	unsigned long	dbcr1;
 #endif
