--- a/src/ap/hw_features.c
+++ b/src/ap/hw_features.c
@@ -218,7 +218,7 @@ static int ieee80211n_allowed_ht40_chann
 	 * 2.4 GHz rules allow all cases where the secondary channel fits into
 	 * the list of allowed channels (already checked above).
 	 */
-	if (iface->current_mode->mode != HOSTAPD_MODE_IEEE80211A)
+	if (1 || iface->current_mode->mode != HOSTAPD_MODE_IEEE80211A)
 		return 1;
 
 	if (iface->conf->secondary_channel > 0)
