--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -2535,6 +2535,8 @@ static int hostapd_config_fill(struct ho
 			}
 		} else if (os_strcmp(buf, "ap_setup_locked") == 0) {
 			bss->ap_setup_locked = atoi(pos);
+		} else if (os_strcmp(buf, "dualband") == 0) {
+			bss->dualband = atoi(pos);
 		} else if (os_strcmp(buf, "uuid") == 0) {
 			if (uuid_str2bin(pos, bss->uuid)) {
 				wpa_printf(MSG_ERROR, "Line %d: invalid UUID",
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -334,6 +334,7 @@ struct hostapd_bss_config {
 	int wps_state;
 #ifdef CONFIG_WPS
 	int ap_setup_locked;
+	int dualband;
 	u8 uuid[16];
 	char *wps_pin_requests;
 	char *device_name;
--- a/src/ap/wps_hostapd.c
+++ b/src/ap/wps_hostapd.c
@@ -957,6 +957,8 @@ int hostapd_init_wps(struct hostapd_data
 	if (conf->ssid.security_policy == SECURITY_STATIC_WEP)
 		cfg.static_wep_only = 1;
 	cfg.dualband = interface_count(hapd->iface) > 1;
+	if (hapd->conf->dualband)
+	    cfg.dualband = 1;
 	if (cfg.dualband)
 		wpa_printf(MSG_DEBUG, "WPS: Dualband AP");
 
