diff -Nru asterisk-1.4.5.org/configure.ac asterisk-1.4.5/configure.ac
--- asterisk-1.4.5.org/configure.ac	2007-06-04 18:02:31.000000000 +0200
+++ asterisk-1.4.5/configure.ac	2007-06-22 09:09:03.000000000 +0200
@@ -419,7 +419,7 @@
    fi
 fi
 
-AST_EXT_LIB_CHECK([IKSEMEL], [iksemel], [iks_start_sasl], [iksemel.h])
+AST_EXT_LIB_CHECK([IKSEMEL], [iksemel], [iks_start_sasl], [iksemel.h], [-lgnutls -lgcrypt -lgpg-error])
 
 if test "${PBX_IKSEMEL}" = 1; then
    AST_EXT_LIB_CHECK([GNUTLS], [gnutls], [gnutls_bye], [gnutls/gnutls.h], [-lz -lgcrypt -lgpg-error])
