diff -Nru asterisk-1.4.5.org/codecs/gsm/Makefile asterisk-1.4.5/codecs/gsm/Makefile
--- asterisk-1.4.5.org/codecs/gsm/Makefile	2007-02-17 01:31:42.000000000 +0100
+++ asterisk-1.4.5/codecs/gsm/Makefile	2007-06-22 09:13:13.000000000 +0200
@@ -38,7 +38,7 @@
 ######### probably require gcc. 
 
 ifeq (, $(findstring $(OSARCH) , Darwin SunOS ))
-ifeq (, $(findstring $(PROC) , x86_64 amd64 ultrasparc sparc64 arm armv5b armeb ppc powerpc ppc64 ia64 s390 bfin mipsel ))
+ifeq (, $(findstring $(PROC) , x86_64 amd64 ultrasparc sparc64 arm armv5b armeb ppc powerpc ppc64 ia64 s390 bfin mipsel mips ))
 ifeq (, $(findstring $(shell uname -m) , ppc ppc64 alpha armv4l s390 ))
 OPTIMIZE+=-march=$(PROC)
 endif
@@ -209,7 +209,7 @@
 # XXX should merge with GSM_OBJECTS
 ifeq ($(OSARCH),linux-gnu)
 ifeq (,$(findstring $(shell uname -m) , x86_64 amd64 ppc ppc64 alpha armv4l sparc64 parisc s390 ))
-ifeq (,$(findstring $(PROC) , arm armv5b armeb powerpc ia64 s390 bfin mipsel ))
+ifeq (,$(findstring $(PROC) , arm armv5b armeb powerpc ia64 s390 bfin mipsel mips ))
 GSM_SOURCES+= $(SRC)/k6opt.s
 endif
 endif
@@ -261,7 +261,7 @@
 
 ifeq ($(OSARCH),linux-gnu)
 ifeq (,$(findstring $(shell uname -m) , x86_64 amd64 ppc ppc64 alpha armv4l sparc64 parisc ))
-ifeq (,$(findstring $(PROC) , arm armv5b armeb powerpc ia64 bfin mipsel ))
+ifeq (,$(findstring $(PROC) , arm armv5b armeb powerpc ia64 bfin mipsel mips ))
 GSM_OBJECTS+= $(SRC)/k6opt.o
 endif
 endif
