diff -puNrb linux-2.6.22.14/kernel/module.c linux/kernel/module.c
--- linux-2.6.22.14/kernel/module.c	2008-01-22 13:55:33.000000000 +0200
+++ linux/kernel/module.c	2008-01-15 10:54:21.000000000 +0200
@@ -1744,6 +1744,10 @@ static struct module *load_module(void _
 	   special cases for the architectures. */
 	layout_sections(mod, hdr, sechdrs, secstrings);
 
+#ifdef __mips
+	module_relayout(hdr, sechdrs, secstrings, symindex, mod);
+#endif
+
 	/* Do the allocs. */
 	ptr = module_alloc(mod->core_size);
 	if (!ptr) {
