diff -puNrb linux-2.6.27.39/drivers/mtd/nand/nand_bbt.c linux/drivers/mtd/nand/nand_bbt.c
--- linux-2.6.27.39/drivers/mtd/nand/nand_bbt.c	2009-11-20 09:54:00.000000000 +0200
+++ linux/drivers/mtd/nand/nand_bbt.c	2009-11-20 09:55:49.000000000 +0200
@@ -328,28 +328,14 @@ static int scan_block_full(struct mtd_in
 static int scan_block_fast(struct mtd_info *mtd, struct nand_bbt_descr *bd,
 			   loff_t offs, uint8_t *buf, int len)
 {
-	struct mtd_oob_ops ops;
+	struct nand_chip *chip = mtd->priv;
 	int j, ret;
 
-	ops.ooblen = mtd->oobsize;
-	ops.oobbuf = buf;
-	ops.ooboffs = 0;
-	ops.datbuf = NULL;
-	ops.mode = MTD_OOB_PLACE;
-
 	for (j = 0; j < len; j++) {
-		/*
-		 * Read the full oob until read_oob is fixed to
-		 * handle single byte reads for 16 bit
-		 * buswidth
-		 */
-		ret = mtd->read_oob(mtd, offs, &ops);
+		ret = chip->block_bad(mtd, offs, 1);
 		if (ret)
 			return ret;
 
-		if (check_short_pattern(buf, bd))
-			return 1;
-
 		offs += mtd->writesize;
 	}
 	return 0;
